{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ae484767-d81e-4a22-956e-7a2dc63f9554",
   "metadata": {},
   "source": [
    "# Looking for associations with microbiome-derived metrics\n",
    "\n",
    "Okay, here is the pitch: suppose you have a novel metric/measure derived on microbiome data and you already calculated this measure for the amplicon sequencing samples in Arivale. What you want to do know is to find feature in the Arivale multiomics data that are associated/correlated with this new measure. This notebook will show you how to do that in a decently efficient manner.\n",
    "\n",
    "For the purpose of having an example we will use alpha diversity as an example. That's really not a novel measure but is one we already have. Let's start by getting the alpha diversity from the Arivale data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b1cac70c-2862-4326-b256-35d3ada1c429",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>public_client_id</th>\n",
       "      <th>vendor_observation_id</th>\n",
       "      <th>vendor_dashboard</th>\n",
       "      <th>client_id</th>\n",
       "      <th>vendor_research</th>\n",
       "      <th>days_in_program</th>\n",
       "      <th>days_since_first_call</th>\n",
       "      <th>days_since_first_draw</th>\n",
       "      <th>month</th>\n",
       "      <th>weekday</th>\n",
       "      <th>...</th>\n",
       "      <th>observed_otus_30000</th>\n",
       "      <th>observed_otus_50000</th>\n",
       "      <th>shannon_100000</th>\n",
       "      <th>shannon_20000</th>\n",
       "      <th>shannon_30000</th>\n",
       "      <th>shannon_50000</th>\n",
       "      <th>simpson_100000</th>\n",
       "      <th>simpson_20000</th>\n",
       "      <th>simpson_30000</th>\n",
       "      <th>simpson_50000</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>01000261</td>\n",
       "      <td>AV15-1592</td>\n",
       "      <td>Second Genome</td>\n",
       "      <td>1966.0</td>\n",
       "      <td>amagis</td>\n",
       "      <td>77.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>Jan</td>\n",
       "      <td>Sun</td>\n",
       "      <td>...</td>\n",
       "      <td>930.0</td>\n",
       "      <td>1133.9</td>\n",
       "      <td>4.287808</td>\n",
       "      <td>4.258003</td>\n",
       "      <td>4.264935</td>\n",
       "      <td>4.280184</td>\n",
       "      <td>0.957</td>\n",
       "      <td>0.957</td>\n",
       "      <td>0.957</td>\n",
       "      <td>0.957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>01001298</td>\n",
       "      <td>22001805511243</td>\n",
       "      <td>research-microbiome</td>\n",
       "      <td>275178.0</td>\n",
       "      <td>amagis</td>\n",
       "      <td>16.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>Sep</td>\n",
       "      <td>Thu</td>\n",
       "      <td>...</td>\n",
       "      <td>494.3</td>\n",
       "      <td>596.4</td>\n",
       "      <td>4.006391</td>\n",
       "      <td>3.986289</td>\n",
       "      <td>3.993914</td>\n",
       "      <td>4.000846</td>\n",
       "      <td>0.954</td>\n",
       "      <td>0.954</td>\n",
       "      <td>0.954</td>\n",
       "      <td>0.954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>01001621</td>\n",
       "      <td>22001612562075</td>\n",
       "      <td>research-microbiome</td>\n",
       "      <td>185773.0</td>\n",
       "      <td>amagis</td>\n",
       "      <td>186.0</td>\n",
       "      <td>144.0</td>\n",
       "      <td>175.0</td>\n",
       "      <td>Jan</td>\n",
       "      <td>Mon</td>\n",
       "      <td>...</td>\n",
       "      <td>453.8</td>\n",
       "      <td>545.2</td>\n",
       "      <td>3.849739</td>\n",
       "      <td>3.832411</td>\n",
       "      <td>3.844887</td>\n",
       "      <td>3.847660</td>\n",
       "      <td>0.955</td>\n",
       "      <td>0.955</td>\n",
       "      <td>0.955</td>\n",
       "      <td>0.955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>01001661</td>\n",
       "      <td>22001612561628</td>\n",
       "      <td>research-microbiome</td>\n",
       "      <td>177541.0</td>\n",
       "      <td>amagis</td>\n",
       "      <td>20.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>-4.0</td>\n",
       "      <td>Jun</td>\n",
       "      <td>Mon</td>\n",
       "      <td>...</td>\n",
       "      <td>453.9</td>\n",
       "      <td>560.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.732598</td>\n",
       "      <td>3.736063</td>\n",
       "      <td>3.740222</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.937</td>\n",
       "      <td>0.937</td>\n",
       "      <td>0.937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>01001798</td>\n",
       "      <td>22001805510925</td>\n",
       "      <td>research-microbiome</td>\n",
       "      <td>278782.0</td>\n",
       "      <td>amagis</td>\n",
       "      <td>12.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>Jan</td>\n",
       "      <td>Mon</td>\n",
       "      <td>...</td>\n",
       "      <td>581.1</td>\n",
       "      <td>720.5</td>\n",
       "      <td>3.835876</td>\n",
       "      <td>3.814389</td>\n",
       "      <td>3.826866</td>\n",
       "      <td>3.833797</td>\n",
       "      <td>0.951</td>\n",
       "      <td>0.951</td>\n",
       "      <td>0.951</td>\n",
       "      <td>0.951</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  public_client_id vendor_observation_id     vendor_dashboard  client_id  \\\n",
       "0         01000261             AV15-1592        Second Genome     1966.0   \n",
       "1         01001298        22001805511243  research-microbiome   275178.0   \n",
       "2         01001621        22001612562075  research-microbiome   185773.0   \n",
       "3         01001661        22001612561628  research-microbiome   177541.0   \n",
       "4         01001798        22001805510925  research-microbiome   278782.0   \n",
       "\n",
       "  vendor_research  days_in_program  days_since_first_call  \\\n",
       "0          amagis             77.0                    3.0   \n",
       "1          amagis             16.0                    9.0   \n",
       "2          amagis            186.0                  144.0   \n",
       "3          amagis             20.0                    6.0   \n",
       "4          amagis             12.0                    7.0   \n",
       "\n",
       "   days_since_first_draw month weekday  ... observed_otus_30000  \\\n",
       "0                   12.0   Jan     Sun  ...               930.0   \n",
       "1                    9.0   Sep     Thu  ...               494.3   \n",
       "2                  175.0   Jan     Mon  ...               453.8   \n",
       "3                   -4.0   Jun     Mon  ...               453.9   \n",
       "4                    9.0   Jan     Mon  ...               581.1   \n",
       "\n",
       "   observed_otus_50000  shannon_100000  shannon_20000  shannon_30000  \\\n",
       "0               1133.9        4.287808       4.258003       4.264935   \n",
       "1                596.4        4.006391       3.986289       3.993914   \n",
       "2                545.2        3.849739       3.832411       3.844887   \n",
       "3                560.6             NaN       3.732598       3.736063   \n",
       "4                720.5        3.835876       3.814389       3.826866   \n",
       "\n",
       "   shannon_50000  simpson_100000  simpson_20000  simpson_30000  simpson_50000  \n",
       "0       4.280184           0.957          0.957          0.957          0.957  \n",
       "1       4.000846           0.954          0.954          0.954          0.954  \n",
       "2       3.847660           0.955          0.955          0.955          0.955  \n",
       "3       3.740222             NaN          0.937          0.937          0.937  \n",
       "4       3.833797           0.951          0.951          0.951          0.951  \n",
       "\n",
       "[5 rows x 37 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import warnings\n",
    "from arivale_data_interface import get_snapshot\n",
    "\n",
    "warnings.simplefilter(\"ignore\")\n",
    "\n",
    "div = get_snapshot(\"microbiome_diversity\")\n",
    "div.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42ccf5a5-7181-4b7e-a43e-7ed310d78fab",
   "metadata": {},
   "source": [
    "The minimum information you will need to keep around is the measure, the public_client_id, vendor_observation_id and days_in_program. If your data only has the vendor_observation_id than you can merge with the the columns in this table to get those other ones. So let's assemble our primary table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e6967d26-c9c5-46ad-b59e-c474d6e2808f",
   "metadata": {},
   "outputs": [],
   "source": [
    "measure = div[[\"diversity_shannon\", \"vendor_observation_id\", \"public_client_id\", \"days_in_program\", \"vendor_dashboard\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7b5b5e5-bb93-4d48-b129-bda481ce192f",
   "metadata": {},
   "source": [
    "Okay that is what we start with. Let's continue."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ec1d105-d5f0-40e9-aee9-4cdba01e885c",
   "metadata": {},
   "source": [
    "## Adding in the confounders\n",
    "\n",
    "The next thing is the confounders, e.g. covariates we want to correct for since they may be associated with our measure and the omics data without having a direct measure <-> omics data connection. For the microbiome data the principal ones are sex, age and BMI, but you could also correct for additional ones like genetic ancestry for instance. We find those measures in the `clients` and `weights` table and will add those in as well. For the weights we encounter a common challenge. There are multiple measurement points for each individual, so which one do we use for the BMI? We will use the closest data point to the microbiome sample. That is why we kept the `days_in_program` data around."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "517e7f28-a129-479b-8333-844288f630ae",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'measure' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[4], line 4\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mpandas\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mpd\u001b[39;00m\n\u001b[1;32m      3\u001b[0m clients \u001b[38;5;241m=\u001b[39m get_snapshot(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mclients\u001b[39m\u001b[38;5;124m\"\u001b[39m)[[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpublic_client_id\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msex\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mage\u001b[39m\u001b[38;5;124m\"\u001b[39m]]\n\u001b[0;32m----> 4\u001b[0m measure \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mmerge(\u001b[43mmeasure\u001b[49m, clients, on\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpublic_client_id\u001b[39m\u001b[38;5;124m\"\u001b[39m, how\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124minner\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      6\u001b[0m weights \u001b[38;5;241m=\u001b[39m get_snapshot(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mweight\u001b[39m\u001b[38;5;124m\"\u001b[39m)[[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mBMI_CALC\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpublic_client_id\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdays_in_program\u001b[39m\u001b[38;5;124m\"\u001b[39m]]\u001b[38;5;241m.\u001b[39mdropna()\n\u001b[1;32m      7\u001b[0m measure \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mmerge_asof(\n\u001b[1;32m      8\u001b[0m     measure[measure\u001b[38;5;241m.\u001b[39mpublic_client_id\u001b[38;5;241m.\u001b[39misin(weights\u001b[38;5;241m.\u001b[39mpublic_client_id)]\u001b[38;5;241m.\u001b[39msort_values(by\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdays_in_program\u001b[39m\u001b[38;5;124m\"\u001b[39m), \n\u001b[1;32m      9\u001b[0m     weights\u001b[38;5;241m.\u001b[39msort_values(by\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdays_in_program\u001b[39m\u001b[38;5;124m\"\u001b[39m), \n\u001b[1;32m     10\u001b[0m     by\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpublic_client_id\u001b[39m\u001b[38;5;124m\"\u001b[39m, on\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdays_in_program\u001b[39m\u001b[38;5;124m\"\u001b[39m, direction\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnearest\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'measure' is not defined"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "clients = get_snapshot(\"clients\")[[\"public_client_id\", \"sex\", \"age\"]]\n",
    "measure = pd.merge(measure, clients, on=\"public_client_id\", how=\"inner\")\n",
    "\n",
    "weights = get_snapshot(\"weight\")[[\"BMI_CALC\", \"public_client_id\", \"days_in_program\"]].dropna()\n",
    "measure = pd.merge_asof(\n",
    "    measure[measure.public_client_id.isin(weights.public_client_id)].sort_values(by=\"days_in_program\"), \n",
    "    weights.sort_values(by=\"days_in_program\"), \n",
    "    by=\"public_client_id\", on=\"days_in_program\", direction=\"nearest\")\n",
    "\n",
    "measure.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f9cb1119-2e30-4007-8515-084d591af700",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>diversity_shannon</th>\n",
       "      <th>vendor_observation_id</th>\n",
       "      <th>public_client_id</th>\n",
       "      <th>days_in_program</th>\n",
       "      <th>vendor_dashboard</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>BMI_CALC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3.870313</td>\n",
       "      <td>AV15-4372</td>\n",
       "      <td>01602320</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>Second Genome</td>\n",
       "      <td>M</td>\n",
       "      <td>48.0</td>\n",
       "      <td>37.229730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4.281822</td>\n",
       "      <td>AV15-3716</td>\n",
       "      <td>01228476</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Second Genome</td>\n",
       "      <td>F</td>\n",
       "      <td>51.0</td>\n",
       "      <td>22.462722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3.318960</td>\n",
       "      <td>AV15-3588</td>\n",
       "      <td>01949838</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Second Genome</td>\n",
       "      <td>F</td>\n",
       "      <td>32.0</td>\n",
       "      <td>25.328546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3.683513</td>\n",
       "      <td>AV15-3840</td>\n",
       "      <td>01602518</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Second Genome</td>\n",
       "      <td>F</td>\n",
       "      <td>50.0</td>\n",
       "      <td>34.154541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4.009163</td>\n",
       "      <td>22001612562257</td>\n",
       "      <td>01360278</td>\n",
       "      <td>1.0</td>\n",
       "      <td>research-microbiome</td>\n",
       "      <td>F</td>\n",
       "      <td>43.0</td>\n",
       "      <td>20.938965</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   diversity_shannon vendor_observation_id public_client_id  days_in_program  \\\n",
       "0           3.870313             AV15-4372         01602320             -2.0   \n",
       "1           4.281822             AV15-3716         01228476              0.0   \n",
       "2           3.318960             AV15-3588         01949838              1.0   \n",
       "3           3.683513             AV15-3840         01602518              1.0   \n",
       "4           4.009163        22001612562257         01360278              1.0   \n",
       "\n",
       "      vendor_dashboard sex   age   BMI_CALC  \n",
       "0        Second Genome   M  48.0  37.229730  \n",
       "1        Second Genome   F  51.0  22.462722  \n",
       "2        Second Genome   F  32.0  25.328546  \n",
       "3        Second Genome   F  50.0  34.154541  \n",
       "4  research-microbiome   F  43.0  20.938965  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "measure.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c335f90-2560-434b-a0c2-4144475b28e6",
   "metadata": {},
   "source": [
    "Okay, some individuals were dropped due to missing data. No biggie. But at least we are now ready to actually start with our association analysis. We will start with the metabolomics data to illustrate the steps for the common omics analyses."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67ea0e16-ed10-4d07-b455-84c3e476970d",
   "metadata": {},
   "source": [
    "## Metabolome\n",
    "\n",
    "Okay each of those analysis will essentially use the same steps which are:\n",
    "1. identify metadata and actual data columns\n",
    "2. merge the omics data with the measure\n",
    "3. transform or normalize and filter features\n",
    "4. go through each feature and merge it with the metadata\n",
    "5. run the regression for each feature and track the results\n",
    "6. merge the results with the feature annotations\n",
    "\n",
    "Let's go then.\n",
    "\n",
    "### 1. Identify metadata and actual data columns\n",
    "\n",
    "We do this so we know which columns denote metadata and which ones actual measures (metabolite abundances)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a0b43a4a-ca19-41dd-8c3f-aae2951c6189",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'measure' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[3], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m metabolites \u001b[38;5;241m=\u001b[39m get_snapshot(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmetabolomics_corrected\u001b[39m\u001b[38;5;124m\"\u001b[39m)\u001b[38;5;241m.\u001b[39msort_values(by\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdays_in_program\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m----> 2\u001b[0m measure \u001b[38;5;241m=\u001b[39m \u001b[43mmeasure\u001b[49m\u001b[38;5;241m.\u001b[39msort_values(by\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdays_in_program\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      3\u001b[0m metabolite_features \u001b[38;5;241m=\u001b[39m metabolites\u001b[38;5;241m.\u001b[39mcolumns[\u001b[38;5;241m8\u001b[39m:]\n\u001b[1;32m      4\u001b[0m metabolites\u001b[38;5;241m.\u001b[39mdays_in_program \u001b[38;5;241m=\u001b[39m metabolites\u001b[38;5;241m.\u001b[39mdays_in_program\u001b[38;5;241m.\u001b[39mastype(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfloat64\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'measure' is not defined"
     ]
    }
   ],
   "source": [
    "metabolites = get_snapshot(\"metabolomics_corrected\").sort_values(by=\"days_in_program\")\n",
    "measure = measure.sort_values(by=\"days_in_program\")\n",
    "metabolite_features = metabolites.columns[8:]\n",
    "metabolites.days_in_program = metabolites.days_in_program.astype(\"float64\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ae79bf1b-df6e-486e-8bb2-5798284ada4b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['35', '50', '55', '62', '71', '93', '98', '111', '112', '117',\n",
       "       ...\n",
       "       '999953114', '999953157', '999953172', '999953266', '999953267',\n",
       "       '999954831', '999954832', '999954834', '999954839', '999954840'],\n",
       "      dtype='object', length=1296)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metabolite_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0c7082fe-8fdf-429a-b56f-ff4a7f3c177c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['public_client_id', 'sample_id', 'days_in_program',\n",
       "       'days_since_first_call', 'days_since_first_draw', 'month', 'weekday',\n",
       "       'season', '35', '50',\n",
       "       ...\n",
       "       '999953114', '999953157', '999953172', '999953266', '999953267',\n",
       "       '999954831', '999954832', '999954834', '999954839', '999954840'],\n",
       "      dtype='object', length=1304)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metabolites.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3879e15-2b7c-44a9-b99b-b1b0a56eb64f",
   "metadata": {},
   "source": [
    "### 2. Merge metadata and measure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1a41a0cb-758f-44c1-be57-315f5b346351",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>diversity_shannon</th>\n",
       "      <th>vendor_observation_id</th>\n",
       "      <th>public_client_id</th>\n",
       "      <th>days_in_program</th>\n",
       "      <th>vendor_dashboard</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>BMI_CALC</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>days_since_first_call</th>\n",
       "      <th>...</th>\n",
       "      <th>999953114</th>\n",
       "      <th>999953157</th>\n",
       "      <th>999953172</th>\n",
       "      <th>999953266</th>\n",
       "      <th>999953267</th>\n",
       "      <th>999954831</th>\n",
       "      <th>999954832</th>\n",
       "      <th>999954834</th>\n",
       "      <th>999954839</th>\n",
       "      <th>999954840</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4.278019</td>\n",
       "      <td>AV15-1348</td>\n",
       "      <td>01306924</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Second Genome</td>\n",
       "      <td>M</td>\n",
       "      <td>56.0</td>\n",
       "      <td>26.541837</td>\n",
       "      <td>A422AU094-002</td>\n",
       "      <td>-155.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.528782</td>\n",
       "      <td>1.606430</td>\n",
       "      <td>1.125981</td>\n",
       "      <td>0.896099</td>\n",
       "      <td>0.838319</td>\n",
       "      <td>1.023423</td>\n",
       "      <td>1.075749</td>\n",
       "      <td>1.051521</td>\n",
       "      <td>0.766951</td>\n",
       "      <td>1.026390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>4.283908</td>\n",
       "      <td>AV15-1295</td>\n",
       "      <td>01421882</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Second Genome</td>\n",
       "      <td>F</td>\n",
       "      <td>19.0</td>\n",
       "      <td>23.793846</td>\n",
       "      <td>A752AT612-002</td>\n",
       "      <td>-5.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.147181</td>\n",
       "      <td>0.758275</td>\n",
       "      <td>1.814588</td>\n",
       "      <td>3.996297</td>\n",
       "      <td>0.935515</td>\n",
       "      <td>0.915287</td>\n",
       "      <td>0.867081</td>\n",
       "      <td>0.871120</td>\n",
       "      <td>1.224561</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>4.403993</td>\n",
       "      <td>AV15-1094</td>\n",
       "      <td>01789546</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Second Genome</td>\n",
       "      <td>F</td>\n",
       "      <td>87.0</td>\n",
       "      <td>18.692937</td>\n",
       "      <td>A976AR221-002</td>\n",
       "      <td>-6.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.153663</td>\n",
       "      <td>0.658271</td>\n",
       "      <td>0.338122</td>\n",
       "      <td>1.162175</td>\n",
       "      <td>1.845431</td>\n",
       "      <td>0.963192</td>\n",
       "      <td>0.942735</td>\n",
       "      <td>0.961368</td>\n",
       "      <td>0.650608</td>\n",
       "      <td>1.703106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>4.542282</td>\n",
       "      <td>AV15-3636</td>\n",
       "      <td>01447446</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Second Genome</td>\n",
       "      <td>F</td>\n",
       "      <td>56.0</td>\n",
       "      <td>21.950816</td>\n",
       "      <td>A904AZ485-002</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.522402</td>\n",
       "      <td>1.073743</td>\n",
       "      <td>1.063270</td>\n",
       "      <td>1.146081</td>\n",
       "      <td>0.942299</td>\n",
       "      <td>1.696602</td>\n",
       "      <td>1.564450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>4.009163</td>\n",
       "      <td>22001612562257</td>\n",
       "      <td>01360278</td>\n",
       "      <td>1.0</td>\n",
       "      <td>research-microbiome</td>\n",
       "      <td>F</td>\n",
       "      <td>43.0</td>\n",
       "      <td>20.938965</td>\n",
       "      <td>A089BI756-002</td>\n",
       "      <td>24.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.260702</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.391536</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.055905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4468</th>\n",
       "      <td>3.684504</td>\n",
       "      <td>AV15-2674</td>\n",
       "      <td>01801678</td>\n",
       "      <td>453.0</td>\n",
       "      <td>Second Genome</td>\n",
       "      <td>M</td>\n",
       "      <td>35.0</td>\n",
       "      <td>20.705900</td>\n",
       "      <td>A706BA371-002</td>\n",
       "      <td>384.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.514679</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.768327</td>\n",
       "      <td>1.997009</td>\n",
       "      <td>1.036581</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.570969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4486</th>\n",
       "      <td>2.850451</td>\n",
       "      <td>AV15-2217</td>\n",
       "      <td>01874590</td>\n",
       "      <td>463.0</td>\n",
       "      <td>Second Genome</td>\n",
       "      <td>M</td>\n",
       "      <td>36.0</td>\n",
       "      <td>21.558261</td>\n",
       "      <td>A928BB321-002</td>\n",
       "      <td>384.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.382292</td>\n",
       "      <td>0.963556</td>\n",
       "      <td>1.770288</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.829983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4551</th>\n",
       "      <td>3.914202</td>\n",
       "      <td>22001805511808</td>\n",
       "      <td>01941857</td>\n",
       "      <td>509.0</td>\n",
       "      <td>research-microbiome</td>\n",
       "      <td>M</td>\n",
       "      <td>54.0</td>\n",
       "      <td>22.148708</td>\n",
       "      <td>A143BO102-003</td>\n",
       "      <td>488.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.016756</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.035061</td>\n",
       "      <td>0.566952</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.032233</td>\n",
       "      <td>1.101631</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4564</th>\n",
       "      <td>4.352964</td>\n",
       "      <td>22001612561459</td>\n",
       "      <td>01524770</td>\n",
       "      <td>517.0</td>\n",
       "      <td>research-microbiome</td>\n",
       "      <td>F</td>\n",
       "      <td>53.0</td>\n",
       "      <td>21.797160</td>\n",
       "      <td>A404BO077-008</td>\n",
       "      <td>531.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.649832</td>\n",
       "      <td>1.048798</td>\n",
       "      <td>3.931804</td>\n",
       "      <td>1.391538</td>\n",
       "      <td>1.358102</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.490642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4715</th>\n",
       "      <td>3.891328</td>\n",
       "      <td>22001805510258</td>\n",
       "      <td>01952180</td>\n",
       "      <td>696.0</td>\n",
       "      <td>research-microbiome</td>\n",
       "      <td>M</td>\n",
       "      <td>55.0</td>\n",
       "      <td>29.257965</td>\n",
       "      <td>A384BN840-008</td>\n",
       "      <td>671.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.733094</td>\n",
       "      <td>1.827141</td>\n",
       "      <td>0.985942</td>\n",
       "      <td>1.285743</td>\n",
       "      <td>0.628764</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.687353</td>\n",
       "      <td>1.597696</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1778 rows Ã— 1310 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      diversity_shannon vendor_observation_id public_client_id  \\\n",
       "3              4.278019             AV15-1348         01306924   \n",
       "6              4.283908             AV15-1295         01421882   \n",
       "7              4.403993             AV15-1094         01789546   \n",
       "8              4.542282             AV15-3636         01447446   \n",
       "9              4.009163        22001612562257         01360278   \n",
       "...                 ...                   ...              ...   \n",
       "4468           3.684504             AV15-2674         01801678   \n",
       "4486           2.850451             AV15-2217         01874590   \n",
       "4551           3.914202        22001805511808         01941857   \n",
       "4564           4.352964        22001612561459         01524770   \n",
       "4715           3.891328        22001805510258         01952180   \n",
       "\n",
       "      days_in_program     vendor_dashboard sex   age   BMI_CALC  \\\n",
       "3                 1.0        Second Genome   M  56.0  26.541837   \n",
       "6                 1.0        Second Genome   F  19.0  23.793846   \n",
       "7                 1.0        Second Genome   F  87.0  18.692937   \n",
       "8                 1.0        Second Genome   F  56.0  21.950816   \n",
       "9                 1.0  research-microbiome   F  43.0  20.938965   \n",
       "...               ...                  ...  ..   ...        ...   \n",
       "4468            453.0        Second Genome   M  35.0  20.705900   \n",
       "4486            463.0        Second Genome   M  36.0  21.558261   \n",
       "4551            509.0  research-microbiome   M  54.0  22.148708   \n",
       "4564            517.0  research-microbiome   F  53.0  21.797160   \n",
       "4715            696.0  research-microbiome   M  55.0  29.257965   \n",
       "\n",
       "          sample_id  days_since_first_call  ...  999953114 999953157  \\\n",
       "3     A422AU094-002                 -155.0  ...   1.528782  1.606430   \n",
       "6     A752AT612-002                   -5.0  ...   1.147181  0.758275   \n",
       "7     A976AR221-002                   -6.0  ...   1.153663  0.658271   \n",
       "8     A904AZ485-002                    1.0  ...        NaN       NaN   \n",
       "9     A089BI756-002                   24.0  ...        NaN  1.260702   \n",
       "...             ...                    ...  ...        ...       ...   \n",
       "4468  A706BA371-002                  384.0  ...   0.514679       NaN   \n",
       "4486  A928BB321-002                  384.0  ...        NaN       NaN   \n",
       "4551  A143BO102-003                  488.0  ...        NaN  1.016756   \n",
       "4564  A404BO077-008                  531.0  ...   1.649832  1.048798   \n",
       "4715  A384BN840-008                  671.0  ...   1.733094  1.827141   \n",
       "\n",
       "     999953172 999953266  999953267  999954831  999954832  999954834  \\\n",
       "3     1.125981  0.896099   0.838319   1.023423   1.075749   1.051521   \n",
       "6     1.814588  3.996297   0.935515   0.915287   0.867081   0.871120   \n",
       "7     0.338122  1.162175   1.845431   0.963192   0.942735   0.961368   \n",
       "8          NaN  1.522402   1.073743   1.063270   1.146081   0.942299   \n",
       "9          NaN       NaN   1.391536        NaN        NaN        NaN   \n",
       "...        ...       ...        ...        ...        ...        ...   \n",
       "4468  0.768327  1.997009   1.036581        NaN        NaN        NaN   \n",
       "4486  1.382292  0.963556   1.770288        NaN        NaN        NaN   \n",
       "4551       NaN  1.035061   0.566952        NaN        NaN        NaN   \n",
       "4564  3.931804  1.391538   1.358102        NaN        NaN        NaN   \n",
       "4715  0.985942  1.285743   0.628764        NaN        NaN        NaN   \n",
       "\n",
       "      999954839  999954840  \n",
       "3      0.766951   1.026390  \n",
       "6      1.224561        NaN  \n",
       "7      0.650608   1.703106  \n",
       "8      1.696602   1.564450  \n",
       "9           NaN   1.055905  \n",
       "...         ...        ...  \n",
       "4468        NaN   0.570969  \n",
       "4486        NaN   0.829983  \n",
       "4551   1.032233   1.101631  \n",
       "4564        NaN   0.490642  \n",
       "4715   0.687353   1.597696  \n",
       "\n",
       "[1778 rows x 1310 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metabolites_merged = pd.merge_asof(\n",
    "    measure, metabolites, \n",
    "    by=\"public_client_id\", \n",
    "    on=\"days_in_program\", \n",
    "    direction=\"nearest\", \n",
    "    tolerance=30.0).dropna(subset=metabolite_features, how=\"all\")\n",
    "metabolite_features = metabolites.columns[metabolites.columns.isin(metabolite_features)]\n",
    "metabolites_merged"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c024818d-1fc8-46af-aca3-bd2403db74fb",
   "metadata": {},
   "source": [
    "Okay we reduced the data set quite by a bit because we requested that we only want samples where blood and fecal samples were talken within 30 days. You can adjust this cutoof however you see fit. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1a3f0a4-e09d-4feb-937d-4ae6bf1d9b73",
   "metadata": {},
   "source": [
    "### 3. Preprocess data\n",
    "\n",
    "This depends a bit on your input but we usually wan to transform our features in a ways that they are comparable and not too rare. So here let's consider only metabolites that appear in at least 75% of all samples. The values are also all positive so we will log-transform and standardize them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2435d101-ff86-4707-b3ba-1c758223fa50",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1778, 929)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bad = metabolites_merged[metabolite_features].isnull().sum() / metabolites_merged.shape[0] > 0.25\n",
    "metabolite_features = bad[~bad].index\n",
    "metabolites_merged = metabolites_merged.drop(columns=bad[bad].index)\n",
    "metabolites_merged.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "557a8953-88b5-42dd-ba6d-eba712ea09f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "metabolites_merged[metabolite_features] = np.log(metabolites_merged[metabolite_features])\n",
    "metabolites_merged[metabolite_features] = metabolites_merged[metabolite_features].apply(lambda x: (x - x.mean()) / x.std())\n",
    "metabolites_merged.rename(columns=dict(zip(metabolite_features, \"metabolite_\" + metabolite_features)), inplace=True)\n",
    "metabolite_features = \"metabolite_\" + metabolite_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "aeb01e4a-da1c-4fd5-8ef7-fdcbf5741ffe",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert (metabolites_merged[metabolite_features].mean().abs() < 1e-6).all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "3d0f4e3f-98f7-4e64-bd3e-b2aca015f4c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>diversity_shannon</th>\n",
       "      <th>vendor_observation_id</th>\n",
       "      <th>public_client_id</th>\n",
       "      <th>days_in_program</th>\n",
       "      <th>vendor_dashboard</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>BMI_CALC</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>days_since_first_call</th>\n",
       "      <th>...</th>\n",
       "      <th>metabolite_999952524</th>\n",
       "      <th>metabolite_999952533</th>\n",
       "      <th>metabolite_999952640</th>\n",
       "      <th>metabolite_999952843</th>\n",
       "      <th>metabolite_999952865</th>\n",
       "      <th>metabolite_999952867</th>\n",
       "      <th>metabolite_999952877</th>\n",
       "      <th>metabolite_999952909</th>\n",
       "      <th>metabolite_999953157</th>\n",
       "      <th>metabolite_999953267</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4.278019</td>\n",
       "      <td>AV15-1348</td>\n",
       "      <td>01306924</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Second Genome</td>\n",
       "      <td>M</td>\n",
       "      <td>56.0</td>\n",
       "      <td>26.541837</td>\n",
       "      <td>A422AU094-002</td>\n",
       "      <td>-155.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.068655</td>\n",
       "      <td>0.677474</td>\n",
       "      <td>-0.560102</td>\n",
       "      <td>-1.702583</td>\n",
       "      <td>1.092412</td>\n",
       "      <td>0.495039</td>\n",
       "      <td>0.654090</td>\n",
       "      <td>0.672189</td>\n",
       "      <td>1.055484</td>\n",
       "      <td>-0.466214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>4.283908</td>\n",
       "      <td>AV15-1295</td>\n",
       "      <td>01421882</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Second Genome</td>\n",
       "      <td>F</td>\n",
       "      <td>19.0</td>\n",
       "      <td>23.793846</td>\n",
       "      <td>A752AT612-002</td>\n",
       "      <td>-5.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.924359</td>\n",
       "      <td>0.402097</td>\n",
       "      <td>0.146529</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.412785</td>\n",
       "      <td>-1.952615</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.114193</td>\n",
       "      <td>-0.565776</td>\n",
       "      <td>-0.216730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>4.403993</td>\n",
       "      <td>AV15-1094</td>\n",
       "      <td>01789546</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Second Genome</td>\n",
       "      <td>F</td>\n",
       "      <td>87.0</td>\n",
       "      <td>18.692937</td>\n",
       "      <td>A976AR221-002</td>\n",
       "      <td>-6.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.629504</td>\n",
       "      <td>-0.871208</td>\n",
       "      <td>1.328348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>4.542282</td>\n",
       "      <td>AV15-3636</td>\n",
       "      <td>01447446</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Second Genome</td>\n",
       "      <td>F</td>\n",
       "      <td>56.0</td>\n",
       "      <td>21.950816</td>\n",
       "      <td>A904AZ485-002</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.096684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>4.009163</td>\n",
       "      <td>22001612562257</td>\n",
       "      <td>01360278</td>\n",
       "      <td>1.0</td>\n",
       "      <td>research-microbiome</td>\n",
       "      <td>F</td>\n",
       "      <td>43.0</td>\n",
       "      <td>20.938965</td>\n",
       "      <td>A089BI756-002</td>\n",
       "      <td>24.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.144189</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.069903</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.315495</td>\n",
       "      <td>-0.467529</td>\n",
       "      <td>0.837372</td>\n",
       "      <td>0.282680</td>\n",
       "      <td>0.532115</td>\n",
       "      <td>0.686309</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 929 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   diversity_shannon vendor_observation_id public_client_id  days_in_program  \\\n",
       "3           4.278019             AV15-1348         01306924              1.0   \n",
       "6           4.283908             AV15-1295         01421882              1.0   \n",
       "7           4.403993             AV15-1094         01789546              1.0   \n",
       "8           4.542282             AV15-3636         01447446              1.0   \n",
       "9           4.009163        22001612562257         01360278              1.0   \n",
       "\n",
       "      vendor_dashboard sex   age   BMI_CALC      sample_id  \\\n",
       "3        Second Genome   M  56.0  26.541837  A422AU094-002   \n",
       "6        Second Genome   F  19.0  23.793846  A752AT612-002   \n",
       "7        Second Genome   F  87.0  18.692937  A976AR221-002   \n",
       "8        Second Genome   F  56.0  21.950816  A904AZ485-002   \n",
       "9  research-microbiome   F  43.0  20.938965  A089BI756-002   \n",
       "\n",
       "   days_since_first_call  ...  metabolite_999952524 metabolite_999952533  \\\n",
       "3                 -155.0  ...             -1.068655             0.677474   \n",
       "6                   -5.0  ...             -0.924359             0.402097   \n",
       "7                   -6.0  ...                   NaN                  NaN   \n",
       "8                    1.0  ...                   NaN                  NaN   \n",
       "9                   24.0  ...              0.144189                  NaN   \n",
       "\n",
       "  metabolite_999952640 metabolite_999952843  metabolite_999952865  \\\n",
       "3            -0.560102            -1.702583              1.092412   \n",
       "6             0.146529                  NaN              0.412785   \n",
       "7                  NaN                  NaN                   NaN   \n",
       "8                  NaN                  NaN                   NaN   \n",
       "9            -0.069903                  NaN              0.315495   \n",
       "\n",
       "   metabolite_999952867  metabolite_999952877  metabolite_999952909  \\\n",
       "3              0.495039              0.654090              0.672189   \n",
       "6             -1.952615                   NaN              0.114193   \n",
       "7                   NaN                   NaN             -1.629504   \n",
       "8                   NaN                   NaN                   NaN   \n",
       "9             -0.467529              0.837372              0.282680   \n",
       "\n",
       "   metabolite_999953157  metabolite_999953267  \n",
       "3              1.055484             -0.466214  \n",
       "6             -0.565776             -0.216730  \n",
       "7             -0.871208              1.328348  \n",
       "8                   NaN              0.096684  \n",
       "9              0.532115              0.686309  \n",
       "\n",
       "[5 rows x 929 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metabolites_merged.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a685acd1-a698-40a2-8286-e0886c5a64d8",
   "metadata": {},
   "source": [
    "### 4. Run the regressions\n",
    "\n",
    "Now we have a clean data set and can start to run the regressions. You will see that this will surprisingly make use of all cores of the machine. This is because conda is setup by default to parallelize basic numeric operations like matrix factorization. Neat!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0941f2c5-b46f-4025-b114-50a82477321b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"></pre>\n"
      ],
      "text/plain": []
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from rich.progress import track\n",
    "from statsmodels.formula.api import ols\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "def get_results(feature):\n",
    "    \"\"\"Get a single association.\n",
    "    \n",
    "    As long as `args` and `df` are assigned this can be used on\n",
    "    any data set.\n",
    "    \"\"\"\n",
    "    formula = f\"{feature} ~ C(sex) + age + BMI_CALC + C(vendor_dashboard) + diversity_shannon\"\n",
    "    fitted = ols(formula, data=df).fit()\n",
    "    return pd.DataFrame({\n",
    "        \"feature\": feature,\n",
    "        \"beta\": fitted.params[\"diversity_shannon\"],\n",
    "        \"t_statistic\": fitted.tvalues[\"diversity_shannon\"],\n",
    "        \"p\": fitted.pvalues[\"diversity_shannon\"],\n",
    "        \"n\": fitted.nobs\n",
    "        }, index=[feature])\n",
    "\n",
    "args = metabolite_features\n",
    "df = metabolites_merged\n",
    "results = map(get_results, track(args))\n",
    "results = list(results)\n",
    "tests = pd.concat(results)\n",
    "tests[\"q\"] = multipletests(tests.p, method=\"fdr_bh\")[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb5a0b05-f4f7-462e-ab4b-6eaa3946dd90",
   "metadata": {},
   "source": [
    "### 5. Merge in annotations\n",
    "\n",
    "Finally we merge in the metadata for the metabolites."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f8159e3c-f492-44a3-a34b-a125f649dac2",
   "metadata": {},
   "outputs": [],
   "source": [
    "anns = get_snapshot(\"metabolomics_metadata\").iloc[:, 0:8]\n",
    "anns[\"feature\"] = \"metabolite_\" + anns.CHEMICAL_ID.astype(str)\n",
    "tests = pd.merge(tests, anns, on=\"feature\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2b7fa05a-f54a-47d6-8f53-d0d0a320d118",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>beta</th>\n",
       "      <th>t_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>n</th>\n",
       "      <th>q</th>\n",
       "      <th>CHEMICAL_ID</th>\n",
       "      <th>SUB_PATHWAY</th>\n",
       "      <th>SUPER_PATHWAY</th>\n",
       "      <th>BIOCHEMICAL_NAME</th>\n",
       "      <th>CAS</th>\n",
       "      <th>KEGG</th>\n",
       "      <th>HMDB</th>\n",
       "      <th>PUBCHEM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>420</th>\n",
       "      <td>metabolite_100002253</td>\n",
       "      <td>1.075114</td>\n",
       "      <td>20.635567</td>\n",
       "      <td>6.062476e-84</td>\n",
       "      <td>1583.0</td>\n",
       "      <td>5.547166e-81</td>\n",
       "      <td>100002253</td>\n",
       "      <td>Food Component/Plant</td>\n",
       "      <td>Xenobiotics</td>\n",
       "      <td>cinnamoylglycine</td>\n",
       "      <td>16534-24-0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HMDB11621</td>\n",
       "      <td>709625.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>metabolite_100000010</td>\n",
       "      <td>1.051025</td>\n",
       "      <td>19.046454</td>\n",
       "      <td>8.262365e-73</td>\n",
       "      <td>1549.0</td>\n",
       "      <td>3.780032e-70</td>\n",
       "      <td>100000010</td>\n",
       "      <td>Phenylalanine and Tyrosine Metabolism</td>\n",
       "      <td>Amino Acid</td>\n",
       "      <td>3-phenylpropionate (hydrocinnamate)</td>\n",
       "      <td>501-52-0</td>\n",
       "      <td>C05629</td>\n",
       "      <td>HMDB00764</td>\n",
       "      <td>107.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>425</th>\n",
       "      <td>metabolite_100002488</td>\n",
       "      <td>-0.977719</td>\n",
       "      <td>-18.893243</td>\n",
       "      <td>3.033731e-72</td>\n",
       "      <td>1660.0</td>\n",
       "      <td>9.252880e-70</td>\n",
       "      <td>100002488</td>\n",
       "      <td>Secondary Bile Acid Metabolism</td>\n",
       "      <td>Lipid</td>\n",
       "      <td>isoursodeoxycholate</td>\n",
       "      <td>78919-26-3</td>\n",
       "      <td>C17662</td>\n",
       "      <td>HMDB00686</td>\n",
       "      <td>127601.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>828</th>\n",
       "      <td>metabolite_999946613</td>\n",
       "      <td>0.975420</td>\n",
       "      <td>18.852741</td>\n",
       "      <td>5.855392e-72</td>\n",
       "      <td>1657.0</td>\n",
       "      <td>1.339421e-69</td>\n",
       "      <td>999946613</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X - 12216</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>metabolite_100001315</td>\n",
       "      <td>0.814344</td>\n",
       "      <td>15.440487</td>\n",
       "      <td>2.029910e-50</td>\n",
       "      <td>1698.0</td>\n",
       "      <td>3.714736e-48</td>\n",
       "      <td>100001315</td>\n",
       "      <td>Phenylalanine and Tyrosine Metabolism</td>\n",
       "      <td>Amino Acid</td>\n",
       "      <td>p-cresol sulfate</td>\n",
       "      <td>3233-57-7</td>\n",
       "      <td>C01468</td>\n",
       "      <td>HMDB11635</td>\n",
       "      <td>4615423.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>metabolite_100000014</td>\n",
       "      <td>0.792415</td>\n",
       "      <td>15.171555</td>\n",
       "      <td>8.639278e-49</td>\n",
       "      <td>1668.0</td>\n",
       "      <td>1.317490e-46</td>\n",
       "      <td>100000014</td>\n",
       "      <td>Benzoate Metabolism</td>\n",
       "      <td>Xenobiotics</td>\n",
       "      <td>hippurate</td>\n",
       "      <td>495-69-2</td>\n",
       "      <td>C01586</td>\n",
       "      <td>HMDB00714</td>\n",
       "      <td>464.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395</th>\n",
       "      <td>metabolite_100002021</td>\n",
       "      <td>0.794967</td>\n",
       "      <td>14.664319</td>\n",
       "      <td>1.722709e-45</td>\n",
       "      <td>1449.0</td>\n",
       "      <td>2.251826e-43</td>\n",
       "      <td>100002021</td>\n",
       "      <td>Steroid</td>\n",
       "      <td>Lipid</td>\n",
       "      <td>5alpha-androstan-3beta,17alpha-diol disulfate</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>812</th>\n",
       "      <td>metabolite_999946507</td>\n",
       "      <td>0.880734</td>\n",
       "      <td>13.685530</td>\n",
       "      <td>3.984056e-40</td>\n",
       "      <td>1418.0</td>\n",
       "      <td>4.556764e-38</td>\n",
       "      <td>999946507</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X - 11850</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>metabolite_100001657</td>\n",
       "      <td>0.699066</td>\n",
       "      <td>12.757053</td>\n",
       "      <td>1.226240e-35</td>\n",
       "      <td>1680.0</td>\n",
       "      <td>1.246678e-33</td>\n",
       "      <td>100001657</td>\n",
       "      <td>Secondary Bile Acid Metabolism</td>\n",
       "      <td>Lipid</td>\n",
       "      <td>glycolithocholate sulfate*</td>\n",
       "      <td>15324-64-8</td>\n",
       "      <td>C11301</td>\n",
       "      <td>HMDB02639</td>\n",
       "      <td>72222.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>438</th>\n",
       "      <td>metabolite_100002911</td>\n",
       "      <td>-0.697544</td>\n",
       "      <td>-12.524210</td>\n",
       "      <td>1.882490e-34</td>\n",
       "      <td>1666.0</td>\n",
       "      <td>1.722478e-32</td>\n",
       "      <td>100002911</td>\n",
       "      <td>Secondary Bile Acid Metabolism</td>\n",
       "      <td>Lipid</td>\n",
       "      <td>glycoursodeoxycholate</td>\n",
       "      <td>64480-66-6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HMDB00708</td>\n",
       "      <td>12310288.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  feature      beta  t_statistic             p       n  \\\n",
       "420  metabolite_100002253  1.075114    20.635567  6.062476e-84  1583.0   \n",
       "167  metabolite_100000010  1.051025    19.046454  8.262365e-73  1549.0   \n",
       "425  metabolite_100002488 -0.977719   -18.893243  3.033731e-72  1660.0   \n",
       "828  metabolite_999946613  0.975420    18.852741  5.855392e-72  1657.0   \n",
       "283  metabolite_100001315  0.814344    15.440487  2.029910e-50  1698.0   \n",
       "168  metabolite_100000014  0.792415    15.171555  8.639278e-49  1668.0   \n",
       "395  metabolite_100002021  0.794967    14.664319  1.722709e-45  1449.0   \n",
       "812  metabolite_999946507  0.880734    13.685530  3.984056e-40  1418.0   \n",
       "355  metabolite_100001657  0.699066    12.757053  1.226240e-35  1680.0   \n",
       "438  metabolite_100002911 -0.697544   -12.524210  1.882490e-34  1666.0   \n",
       "\n",
       "                q  CHEMICAL_ID                            SUB_PATHWAY  \\\n",
       "420  5.547166e-81    100002253                   Food Component/Plant   \n",
       "167  3.780032e-70    100000010  Phenylalanine and Tyrosine Metabolism   \n",
       "425  9.252880e-70    100002488         Secondary Bile Acid Metabolism   \n",
       "828  1.339421e-69    999946613                                    NaN   \n",
       "283  3.714736e-48    100001315  Phenylalanine and Tyrosine Metabolism   \n",
       "168  1.317490e-46    100000014                    Benzoate Metabolism   \n",
       "395  2.251826e-43    100002021                                Steroid   \n",
       "812  4.556764e-38    999946507                                    NaN   \n",
       "355  1.246678e-33    100001657         Secondary Bile Acid Metabolism   \n",
       "438  1.722478e-32    100002911         Secondary Bile Acid Metabolism   \n",
       "\n",
       "    SUPER_PATHWAY                               BIOCHEMICAL_NAME         CAS  \\\n",
       "420   Xenobiotics                               cinnamoylglycine  16534-24-0   \n",
       "167    Amino Acid            3-phenylpropionate (hydrocinnamate)    501-52-0   \n",
       "425         Lipid                            isoursodeoxycholate  78919-26-3   \n",
       "828           NaN                                      X - 12216         NaN   \n",
       "283    Amino Acid                               p-cresol sulfate   3233-57-7   \n",
       "168   Xenobiotics                                      hippurate    495-69-2   \n",
       "395         Lipid  5alpha-androstan-3beta,17alpha-diol disulfate         NaN   \n",
       "812           NaN                                      X - 11850         NaN   \n",
       "355         Lipid                     glycolithocholate sulfate*  15324-64-8   \n",
       "438         Lipid                          glycoursodeoxycholate  64480-66-6   \n",
       "\n",
       "       KEGG       HMDB     PUBCHEM  \n",
       "420     NaN  HMDB11621    709625.0  \n",
       "167  C05629  HMDB00764       107.0  \n",
       "425  C17662  HMDB00686    127601.0  \n",
       "828     NaN        NaN         NaN  \n",
       "283  C01468  HMDB11635   4615423.0  \n",
       "168  C01586  HMDB00714       464.0  \n",
       "395     NaN        NaN         NaN  \n",
       "812     NaN        NaN         NaN  \n",
       "355  C11301  HMDB02639     72222.0  \n",
       "438     NaN  HMDB00708  12310288.0  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# tests.to_csv(\"metabolome_tests.csv\", index=False)  # uncomment to save the result\n",
    "tests.sort_values(by=\"q\").head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5a40826-af01-4fda-a25c-8eea1a1f503c",
   "metadata": {},
   "source": [
    "Okay, that was this. You could do exactly the same for proteomics. Let's run through it with clinical labs here since we don't really want to adjust the variables that much. We will group the steps this time.\n",
    "\n",
    "## Clinical labs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5d3f8b7a-aa3c-456e-ab66-21ac754a541a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find the features\n",
    "chems = get_snapshot(\"chemistries\", clean=True).sort_values(by=\"days_in_program\")\n",
    "chem_features = chems.columns[12:]  # inspect the table to find out where the metadata ends \n",
    "chems.days_in_program = chems.days_in_program.astype(\"float64\")\n",
    "\n",
    "# Merge with measure\n",
    "chems_merged = pd.merge_asof(\n",
    "    measure, chems, \n",
    "    by=\"public_client_id\", \n",
    "    on=\"days_in_program\", \n",
    "    direction=\"nearest\", \n",
    "    tolerance=30.0).dropna(subset=chem_features, how=\"all\")\n",
    "chem_features = chems.columns[chems.columns.isin(chem_features)]\n",
    "\n",
    "# Filter uncomplete\n",
    "bad = chems_merged[chem_features].isnull().sum() / chems_merged.shape[0] > 0.25\n",
    "chem_features = bad[~bad].index\n",
    "chems_merged = chems_merged.drop(columns=bad[bad].index)\n",
    "\n",
    "# We don't transform the values further since some of them are integer etc. and they have associated units we don't want to break\n",
    "# However, that mean we can't compare the beta values between features."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "8aec52ae-b11a-429c-bb13-d12817596a33",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['diversity_shannon', 'vendor_observation_id_x', 'public_client_id',\n",
       "       'days_in_program', 'vendor_dashboard', 'sex', 'age', 'BMI_CALC',\n",
       "       'vendor', 'vendor_observation_id_y', 'observation_id', 'reflexive',\n",
       "       'fasting', 'days_since_first_call', 'days_since_first_draw', 'month',\n",
       "       'weekday', 'season', 'A/G RATIO', 'ALAT (SGPT)', 'ALBUMIN',\n",
       "       'ALKALINE PHOSPHATASE', 'ARACHIDONIC ACID', 'ASAT (SGOT)', 'BASOPHILS',\n",
       "       'BASOPHILS ABSOLUTE', 'BILIRUBIN, TOTAL', 'BUN/CREAT RATIO', 'CALCIUM',\n",
       "       'CARBON DIOXIDE (CO2)', 'CHLORIDE', 'CHOLESTEROL, TOTAL',\n",
       "       'CREATININE ENZ, SER', 'CRP HIGH SENSITIVITY', 'DHA', 'DPA',\n",
       "       'EOSINOPHILS', 'EOSINOPHILS ABSOLUTE', 'EPA', 'FERRITIN', 'GFR, MDRD',\n",
       "       'GFR, MDRD, AFRICAN AM', 'GGT', 'GLOBULIN', 'GLUCOSE',\n",
       "       'GLYCOHEMOGLOBIN A1C', 'HDL CHOL DIRECT', 'HDL PARTICLE NUMBER',\n",
       "       'HEMATOCRIT', 'HEMOGLOBIN', 'HOMA-IR', 'HOMOCYSTEINE, SERUM',\n",
       "       'IMMATURE GRANULOCYTES', 'IMMATURE GRANULOCYTES ABSOLUTE', 'INSULIN',\n",
       "       'LDL PARTICLE NUMBER', 'LDL SMALL', 'LDL-CHOL CALCULATION', 'LDL_SIZE',\n",
       "       'LINOLEIC_ACID', 'LPIR_SCORE', 'LYMPHOCYTES', 'LYMPHOCYTES ABSOLUTE',\n",
       "       'MAGNESIUM, SERUM', 'MCH', 'MCHC', 'MCV', 'MERCURY, BLOOD', 'MONOCYTES',\n",
       "       'MONOCYTES ABSOLUTE', 'OMEGA-3 INDEX', 'OMEGA-6/OMEGA-3 RATIO',\n",
       "       'OMEGA_3_TOTAL', 'OMEGA_6_TOTAL', 'PLATELET COUNT THOUSAND',\n",
       "       'POTASSIUM', 'PROTEIN, TOTAL SERUM', 'RDW', 'RED CELL COUNT', 'SODIUM',\n",
       "       'TOTAL NEUTROPHILS', 'TOTAL NEUTROPHILS AB', 'TRIGLYCERIDES',\n",
       "       'Triglyceride HDL Ratio', 'UREA NITROGEN', 'URIC ACID',\n",
       "       'VITAMIN D, 25-OH TOT', 'WHITE CELL COUNT'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chems_merged.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f14ee2ec-468e-4d2f-a9a7-569585abee78",
   "metadata": {},
   "source": [
    "Looks pretty concise this way and we can start to run the testing loop again."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4969e0e2-6d56-48d6-afff-a2bca028cdf4",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<unknown>, line 1)",
     "output_type": "error",
     "traceback": [
      "Traceback \u001b[0;36m(most recent call last)\u001b[0m:\n",
      "\u001b[0m  File \u001b[1;32m~/miniconda3/envs/arivale-py/lib/python3.11/site-packages/IPython/core/interactiveshell.py:3526\u001b[0m in \u001b[1;35mrun_code\u001b[0m\n    exec(code_obj, self.user_global_ns, self.user_ns)\u001b[0m\n",
      "\u001b[0m  Cell \u001b[1;32mIn[17], line 4\u001b[0m\n    results = list(results)\u001b[0m\n",
      "\u001b[0m  Cell \u001b[1;32mIn[13], line 12\u001b[0m in \u001b[1;35mget_results\u001b[0m\n    fitted = ols(formula, data=df).fit()\u001b[0m\n",
      "\u001b[0m  File \u001b[1;32m~/miniconda3/envs/arivale-py/lib/python3.11/site-packages/statsmodels/base/model.py:203\u001b[0m in \u001b[1;35mfrom_formula\u001b[0m\n    tmp = handle_formula_data(data, None, formula, depth=eval_env,\u001b[0m\n",
      "\u001b[0m  File \u001b[1;32m~/miniconda3/envs/arivale-py/lib/python3.11/site-packages/statsmodels/formula/formulatools.py:63\u001b[0m in \u001b[1;35mhandle_formula_data\u001b[0m\n    result = dmatrices(formula, Y, depth, return_type='dataframe',\u001b[0m\n",
      "\u001b[0m  File \u001b[1;32m~/miniconda3/envs/arivale-py/lib/python3.11/site-packages/patsy/highlevel.py:309\u001b[0m in \u001b[1;35mdmatrices\u001b[0m\n    (lhs, rhs) = _do_highlevel_design(formula_like, data, eval_env,\u001b[0m\n",
      "\u001b[0m  File \u001b[1;32m~/miniconda3/envs/arivale-py/lib/python3.11/site-packages/patsy/highlevel.py:164\u001b[0m in \u001b[1;35m_do_highlevel_design\u001b[0m\n    design_infos = _try_incr_builders(formula_like, data_iter_maker, eval_env,\u001b[0m\n",
      "\u001b[0m  File \u001b[1;32m~/miniconda3/envs/arivale-py/lib/python3.11/site-packages/patsy/highlevel.py:66\u001b[0m in \u001b[1;35m_try_incr_builders\u001b[0m\n    return design_matrix_builders([formula_like.lhs_termlist,\u001b[0m\n",
      "\u001b[0m  File \u001b[1;32m~/miniconda3/envs/arivale-py/lib/python3.11/site-packages/patsy/build.py:689\u001b[0m in \u001b[1;35mdesign_matrix_builders\u001b[0m\n    factor_states = _factors_memorize(all_factors, data_iter_maker, eval_env)\u001b[0m\n",
      "\u001b[0m  File \u001b[1;32m~/miniconda3/envs/arivale-py/lib/python3.11/site-packages/patsy/build.py:354\u001b[0m in \u001b[1;35m_factors_memorize\u001b[0m\n    which_pass = factor.memorize_passes_needed(state, eval_env)\u001b[0m\n",
      "\u001b[0m  File \u001b[1;32m~/miniconda3/envs/arivale-py/lib/python3.11/site-packages/patsy/eval.py:478\u001b[0m in \u001b[1;35mmemorize_passes_needed\u001b[0m\n    subset_names = [name for name in ast_names(self.code)\u001b[0m\n",
      "\u001b[0m  File \u001b[1;32m~/miniconda3/envs/arivale-py/lib/python3.11/site-packages/patsy/eval.py:478\u001b[0m in \u001b[1;35m<listcomp>\u001b[0m\n    subset_names = [name for name in ast_names(self.code)\u001b[0m\n",
      "\u001b[0m  File \u001b[1;32m~/miniconda3/envs/arivale-py/lib/python3.11/site-packages/patsy/eval.py:109\u001b[0m in \u001b[1;35mast_names\u001b[0m\n    for node in ast.walk(ast.parse(code)):\u001b[0m\n",
      "\u001b[0;36m  File \u001b[0;32m~/miniconda3/envs/arivale-py/lib/python3.11/ast.py:50\u001b[0;36m in \u001b[0;35mparse\u001b[0;36m\n\u001b[0;31m    return compile(source, filename, mode, flags,\u001b[0;36m\n",
      "\u001b[0;36m  File \u001b[0;32m<unknown>:1\u001b[0;36m\u001b[0m\n\u001b[0;31m    G RATIO\u001b[0m\n\u001b[0m      ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "args = chem_features\n",
    "df = chems_merged  # don't forget to reassign those\n",
    "results = map(get_results, track(args))\n",
    "results = list(results)\n",
    "tests_chems = pd.concat(results)\n",
    "tests_chems[\"q\"] = multipletests(tests_chems.p, method=\"fdr_bh\")[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ddb4d9b-c636-41ef-a326-4162236f54bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "anns = get_snapshot(\"chemistries_metadata\")\n",
    "anns[\"feature\"] = anns.Name\n",
    "tests_chems = pd.merge(tests_chems, anns, on=\"feature\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "794d1c1f-457d-42a0-a11e-70c145fc5cda",
   "metadata": {},
   "outputs": [],
   "source": [
    "tests_chems.sort_values(by=\"q\").head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd82876f-afa7-442f-8f17-2215966c8917",
   "metadata": {},
   "source": [
    "## Questionnaires\n",
    "\n",
    "Okay same things as before but we get the questionnaire data from the general assessments and will express it on an integer scale from 0-M where 0 is little and M is a lot. For that we will only use the \"enum\" and \"int\" scales."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "53047222-2fc8-4c71-8c07-6a506186fca9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>diversity_shannon</th>\n",
       "      <th>vendor_observation_id</th>\n",
       "      <th>public_client_id</th>\n",
       "      <th>days_in_program</th>\n",
       "      <th>vendor_dashboard</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>BMI_CALC</th>\n",
       "      <th>vendor</th>\n",
       "      <th>days_since_first_call</th>\n",
       "      <th>...</th>\n",
       "      <th>assessment:pss-four-item:insurmountable:enum</th>\n",
       "      <th>assessment:pss-four-item:out_of_control:enum</th>\n",
       "      <th>assessment:vitals:antibiotics:enum</th>\n",
       "      <th>assessment:vitals:blood_pressure:enum</th>\n",
       "      <th>assessment:vitals:blood_sugar:enum</th>\n",
       "      <th>assessment:vitals:cholesterol:enum</th>\n",
       "      <th>assessment:vitals:colonoscopy:enum</th>\n",
       "      <th>assessment:vitals:ethnicity:enum</th>\n",
       "      <th>assessment:vitals:mammogram:enum</th>\n",
       "      <th>assessment:vitals:race:enum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>2.712285</td>\n",
       "      <td>22001805511579</td>\n",
       "      <td>01156594</td>\n",
       "      <td>3.0</td>\n",
       "      <td>research-microbiome</td>\n",
       "      <td>F</td>\n",
       "      <td>72.0</td>\n",
       "      <td>32.220833</td>\n",
       "      <td>Assessments</td>\n",
       "      <td>-9.0</td>\n",
       "      <td>...</td>\n",
       "      <td>(1) Almost Never</td>\n",
       "      <td>(1) Almost Never</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>4.530410</td>\n",
       "      <td>22001803513494</td>\n",
       "      <td>01362239</td>\n",
       "      <td>4.0</td>\n",
       "      <td>research-microbiome</td>\n",
       "      <td>F</td>\n",
       "      <td>70.0</td>\n",
       "      <td>22.594123</td>\n",
       "      <td>Assessments</td>\n",
       "      <td>-4.0</td>\n",
       "      <td>...</td>\n",
       "      <td>(0) Never</td>\n",
       "      <td>(0) Never</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>3.257099</td>\n",
       "      <td>22001805511823</td>\n",
       "      <td>01730310</td>\n",
       "      <td>4.0</td>\n",
       "      <td>research-microbiome</td>\n",
       "      <td>M</td>\n",
       "      <td>51.0</td>\n",
       "      <td>28.693878</td>\n",
       "      <td>Assessments</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>(3) Fairly Often</td>\n",
       "      <td>(1) Almost Never</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>3.562083</td>\n",
       "      <td>22001805520631</td>\n",
       "      <td>01712678</td>\n",
       "      <td>4.0</td>\n",
       "      <td>research-microbiome</td>\n",
       "      <td>F</td>\n",
       "      <td>32.0</td>\n",
       "      <td>42.252038</td>\n",
       "      <td>Assessments</td>\n",
       "      <td>-13.0</td>\n",
       "      <td>...</td>\n",
       "      <td>(2) Sometimes</td>\n",
       "      <td>(2) Sometimes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>3.620308</td>\n",
       "      <td>22001804570150</td>\n",
       "      <td>01225920</td>\n",
       "      <td>4.0</td>\n",
       "      <td>research-microbiome</td>\n",
       "      <td>M</td>\n",
       "      <td>40.0</td>\n",
       "      <td>34.256669</td>\n",
       "      <td>Assessments</td>\n",
       "      <td>-21.0</td>\n",
       "      <td>...</td>\n",
       "      <td>(2) Sometimes</td>\n",
       "      <td>(3) Fairly Often</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5071</th>\n",
       "      <td>3.362457</td>\n",
       "      <td>22001805521425</td>\n",
       "      <td>01994330</td>\n",
       "      <td>1348.0</td>\n",
       "      <td>research-microbiome</td>\n",
       "      <td>M</td>\n",
       "      <td>48.0</td>\n",
       "      <td>28.112085</td>\n",
       "      <td>Assessments</td>\n",
       "      <td>1309.0</td>\n",
       "      <td>...</td>\n",
       "      <td>(1) Almost Never</td>\n",
       "      <td>(1) Almost Never</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5074</th>\n",
       "      <td>3.354139</td>\n",
       "      <td>22001805510143</td>\n",
       "      <td>01108982</td>\n",
       "      <td>1354.0</td>\n",
       "      <td>research-microbiome</td>\n",
       "      <td>M</td>\n",
       "      <td>62.0</td>\n",
       "      <td>22.081310</td>\n",
       "      <td>Assessments</td>\n",
       "      <td>1311.0</td>\n",
       "      <td>...</td>\n",
       "      <td>(1) Almost Never</td>\n",
       "      <td>(0) Never</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5076</th>\n",
       "      <td>3.868454</td>\n",
       "      <td>22001805521073</td>\n",
       "      <td>01441946</td>\n",
       "      <td>1367.0</td>\n",
       "      <td>research-microbiome</td>\n",
       "      <td>F</td>\n",
       "      <td>44.0</td>\n",
       "      <td>19.732234</td>\n",
       "      <td>Assessments</td>\n",
       "      <td>1334.0</td>\n",
       "      <td>...</td>\n",
       "      <td>(2) Sometimes</td>\n",
       "      <td>(2) Sometimes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5078</th>\n",
       "      <td>4.207403</td>\n",
       "      <td>22001805510387</td>\n",
       "      <td>01023946</td>\n",
       "      <td>1371.0</td>\n",
       "      <td>research-microbiome</td>\n",
       "      <td>F</td>\n",
       "      <td>43.0</td>\n",
       "      <td>27.259184</td>\n",
       "      <td>Assessments</td>\n",
       "      <td>1339.0</td>\n",
       "      <td>...</td>\n",
       "      <td>(2) Sometimes</td>\n",
       "      <td>(3) Fairly Often</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5079</th>\n",
       "      <td>4.273946</td>\n",
       "      <td>22001805521617</td>\n",
       "      <td>01483778</td>\n",
       "      <td>1377.0</td>\n",
       "      <td>research-microbiome</td>\n",
       "      <td>F</td>\n",
       "      <td>39.0</td>\n",
       "      <td>23.440253</td>\n",
       "      <td>Assessments</td>\n",
       "      <td>1324.0</td>\n",
       "      <td>...</td>\n",
       "      <td>(2) Sometimes</td>\n",
       "      <td>(3) Fairly Often</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>966 rows Ã— 3496 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      diversity_shannon vendor_observation_id public_client_id  \\\n",
       "45             2.712285        22001805511579         01156594   \n",
       "85             4.530410        22001803513494         01362239   \n",
       "95             3.257099        22001805511823         01730310   \n",
       "116            3.562083        22001805520631         01712678   \n",
       "120            3.620308        22001804570150         01225920   \n",
       "...                 ...                   ...              ...   \n",
       "5071           3.362457        22001805521425         01994330   \n",
       "5074           3.354139        22001805510143         01108982   \n",
       "5076           3.868454        22001805521073         01441946   \n",
       "5078           4.207403        22001805510387         01023946   \n",
       "5079           4.273946        22001805521617         01483778   \n",
       "\n",
       "      days_in_program     vendor_dashboard sex   age   BMI_CALC       vendor  \\\n",
       "45                3.0  research-microbiome   F  72.0  32.220833  Assessments   \n",
       "85                4.0  research-microbiome   F  70.0  22.594123  Assessments   \n",
       "95                4.0  research-microbiome   M  51.0  28.693878  Assessments   \n",
       "116               4.0  research-microbiome   F  32.0  42.252038  Assessments   \n",
       "120               4.0  research-microbiome   M  40.0  34.256669  Assessments   \n",
       "...               ...                  ...  ..   ...        ...          ...   \n",
       "5071           1348.0  research-microbiome   M  48.0  28.112085  Assessments   \n",
       "5074           1354.0  research-microbiome   M  62.0  22.081310  Assessments   \n",
       "5076           1367.0  research-microbiome   F  44.0  19.732234  Assessments   \n",
       "5078           1371.0  research-microbiome   F  43.0  27.259184  Assessments   \n",
       "5079           1377.0  research-microbiome   F  39.0  23.440253  Assessments   \n",
       "\n",
       "      days_since_first_call  ...  \\\n",
       "45                     -9.0  ...   \n",
       "85                     -4.0  ...   \n",
       "95                      0.0  ...   \n",
       "116                   -13.0  ...   \n",
       "120                   -21.0  ...   \n",
       "...                     ...  ...   \n",
       "5071                 1309.0  ...   \n",
       "5074                 1311.0  ...   \n",
       "5076                 1334.0  ...   \n",
       "5078                 1339.0  ...   \n",
       "5079                 1324.0  ...   \n",
       "\n",
       "      assessment:pss-four-item:insurmountable:enum  \\\n",
       "45                                (1) Almost Never   \n",
       "85                                       (0) Never   \n",
       "95                                (3) Fairly Often   \n",
       "116                                  (2) Sometimes   \n",
       "120                                  (2) Sometimes   \n",
       "...                                            ...   \n",
       "5071                              (1) Almost Never   \n",
       "5074                              (1) Almost Never   \n",
       "5076                                 (2) Sometimes   \n",
       "5078                                 (2) Sometimes   \n",
       "5079                                 (2) Sometimes   \n",
       "\n",
       "     assessment:pss-four-item:out_of_control:enum  \\\n",
       "45                               (1) Almost Never   \n",
       "85                                      (0) Never   \n",
       "95                               (1) Almost Never   \n",
       "116                                 (2) Sometimes   \n",
       "120                              (3) Fairly Often   \n",
       "...                                           ...   \n",
       "5071                             (1) Almost Never   \n",
       "5074                                    (0) Never   \n",
       "5076                                (2) Sometimes   \n",
       "5078                             (3) Fairly Often   \n",
       "5079                             (3) Fairly Often   \n",
       "\n",
       "     assessment:vitals:antibiotics:enum assessment:vitals:blood_pressure:enum  \\\n",
       "45                                  NaN                                   NaN   \n",
       "85                                  NaN                                   NaN   \n",
       "95                                  NaN                                   NaN   \n",
       "116                                 NaN                                   NaN   \n",
       "120                                 NaN                                   NaN   \n",
       "...                                 ...                                   ...   \n",
       "5071                                NaN                                   NaN   \n",
       "5074                                NaN                                   NaN   \n",
       "5076                                NaN                                   NaN   \n",
       "5078                                NaN                                   NaN   \n",
       "5079                                NaN                                   NaN   \n",
       "\n",
       "     assessment:vitals:blood_sugar:enum assessment:vitals:cholesterol:enum  \\\n",
       "45                                  NaN                                NaN   \n",
       "85                                  NaN                                NaN   \n",
       "95                                  NaN                                NaN   \n",
       "116                                 NaN                                NaN   \n",
       "120                                 NaN                                NaN   \n",
       "...                                 ...                                ...   \n",
       "5071                                NaN                                NaN   \n",
       "5074                                NaN                                NaN   \n",
       "5076                                NaN                                NaN   \n",
       "5078                                NaN                                NaN   \n",
       "5079                                NaN                                NaN   \n",
       "\n",
       "     assessment:vitals:colonoscopy:enum assessment:vitals:ethnicity:enum  \\\n",
       "45                                  NaN                              NaN   \n",
       "85                                  NaN                              NaN   \n",
       "95                                  NaN                              NaN   \n",
       "116                                 NaN                              NaN   \n",
       "120                                 NaN                              NaN   \n",
       "...                                 ...                              ...   \n",
       "5071                                NaN                              NaN   \n",
       "5074                                NaN                              NaN   \n",
       "5076                                NaN                              NaN   \n",
       "5078                                NaN                              NaN   \n",
       "5079                                NaN                              NaN   \n",
       "\n",
       "     assessment:vitals:mammogram:enum assessment:vitals:race:enum  \n",
       "45                                NaN                         NaN  \n",
       "85                                NaN                         NaN  \n",
       "95                                NaN                         NaN  \n",
       "116                               NaN                         NaN  \n",
       "120                               NaN                         NaN  \n",
       "...                               ...                         ...  \n",
       "5071                              NaN                         NaN  \n",
       "5074                              NaN                         NaN  \n",
       "5076                              NaN                         NaN  \n",
       "5078                              NaN                         NaN  \n",
       "5079                              NaN                         NaN  \n",
       "\n",
       "[966 rows x 3496 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "qs = get_snapshot(\"assessments\", clean=True).sort_values(by=\"days_in_program\")\n",
    "q_features = qs.columns[8:]\n",
    "q_features = q_features[q_features.str.endswith(\"_enum\") | q_features.str.endswith(\"_int\")]\n",
    "qs.days_in_program = qs.days_in_program.astype(\"float64\")\n",
    "\n",
    "# Merge with measure\n",
    "qs_merged = pd.merge_asof(\n",
    "    measure, qs.dropna(subset=\"days_in_program\"), \n",
    "    by=\"public_client_id\", \n",
    "    on=\"days_in_program\", \n",
    "    direction=\"nearest\", \n",
    "    tolerance=30.0).dropna(subset=q_features, how=\"all\")\n",
    "q_features = qs.columns[qs.columns.isin(q_features)]\n",
    "\n",
    "# Filter uncomplete\n",
    "bad = qs_merged[q_features].isnull().sum() / qs_merged.shape[0] > 0.25\n",
    "q_features = bad[~bad].index\n",
    "qs_merged = qs_merged.drop(columns=bad[bad].index)\n",
    "\n",
    "# The data is all categorical so we will make it ordinal now\n",
    "qs_merged[q_features] = qs_merged[q_features].apply(lambda x: x.str.split(\"\\\\(|\\\\)\", regex=True).str[1]).astype(\"float64\")\n",
    "qs_merged"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "407c9a8a-4d10-422b-bebf-d982d1c5d9c6",
   "metadata": {},
   "source": [
    "And we test again."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "0b1982bd-f570-4c9a-bcb3-213c2a88d3ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"></pre>\n"
      ],
      "text/plain": []
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"></pre>\n"
      ],
      "text/plain": []
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "args = q_features\n",
    "df = qs_merged  # don't forget to reassign those\n",
    "results = map(get_results, track(args))\n",
    "results = list(results)\n",
    "tests_qs = pd.concat(results)\n",
    "tests_qs[\"q\"] = multipletests(tests_qs.p, method=\"fdr_bh\")[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a021c813-5b15-45dc-aa03-105d51a5cb7a",
   "metadata": {},
   "source": [
    "For the annotations we will simply append the meaning of the numeric scale here. Makes it easier to interpret the result (like the directon of beta)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f6f5f20a-a283-4ef2-b448-4f0165e46c4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "scales = qs[q_features].apply(lambda x: \" | \".join(x.astype(str).sort_values().unique()) if x.dtype == \"O\" else \"numeric\")\n",
    "tests_qs[\"scales\"] = scales[tests_qs.feature].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "f1a80481-97df-4fa2-8355-d4ad4d7ddd27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>beta</th>\n",
       "      <th>t_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>n</th>\n",
       "      <th>q</th>\n",
       "      <th>scales</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>assessment_health_history_prescription_medication_use_yes_no_int</th>\n",
       "      <td>assessment_health_history_prescription_medicat...</td>\n",
       "      <td>-0.087469</td>\n",
       "      <td>-2.585243</td>\n",
       "      <td>0.009878</td>\n",
       "      <td>964.0</td>\n",
       "      <td>0.079021</td>\n",
       "      <td>(1) No | (2) Yes | nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assessment_health_history_drugs_asthma_int</th>\n",
       "      <td>assessment_health_history_drugs_asthma_int</td>\n",
       "      <td>-0.080982</td>\n",
       "      <td>-1.932916</td>\n",
       "      <td>0.053542</td>\n",
       "      <td>961.0</td>\n",
       "      <td>0.214166</td>\n",
       "      <td>(0) Never | (1) Currently | (2) Formerly | nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assessment_health_history_drugs_allergy_int</th>\n",
       "      <td>assessment_health_history_drugs_allergy_int</td>\n",
       "      <td>-0.061368</td>\n",
       "      <td>-1.214417</td>\n",
       "      <td>0.224889</td>\n",
       "      <td>961.0</td>\n",
       "      <td>0.599703</td>\n",
       "      <td>(0) Never | (1) Currently | (2) Formerly | nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assessment_health_history_drugs_anti_coagulation_int</th>\n",
       "      <td>assessment_health_history_drugs_anti_coagulati...</td>\n",
       "      <td>-0.005262</td>\n",
       "      <td>-0.176631</td>\n",
       "      <td>0.859836</td>\n",
       "      <td>960.0</td>\n",
       "      <td>0.859836</td>\n",
       "      <td>(0) Never | (1) Currently | (2) Formerly | nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assessment_health_history_drugs_osteoporosis_int</th>\n",
       "      <td>assessment_health_history_drugs_osteoporosis_int</td>\n",
       "      <td>-0.004209</td>\n",
       "      <td>-0.215083</td>\n",
       "      <td>0.829749</td>\n",
       "      <td>961.0</td>\n",
       "      <td>0.859836</td>\n",
       "      <td>(0) Never | (1) Currently | (2) Formerly | nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assessment_health_history_drugs_psychiatric_int</th>\n",
       "      <td>assessment_health_history_drugs_psychiatric_int</td>\n",
       "      <td>-0.028410</td>\n",
       "      <td>-0.657168</td>\n",
       "      <td>0.511231</td>\n",
       "      <td>961.0</td>\n",
       "      <td>0.859836</td>\n",
       "      <td>(0) Never | (1) Currently | (2) Formerly | nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assessment_health_history_over_the_counter_medication_use_int</th>\n",
       "      <td>assessment_health_history_over_the_counter_med...</td>\n",
       "      <td>0.024875</td>\n",
       "      <td>0.486631</td>\n",
       "      <td>0.626632</td>\n",
       "      <td>957.0</td>\n",
       "      <td>0.859836</td>\n",
       "      <td>(0) No | (1) Rarely/Occasionally | (2) Yes | nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assessment_health_history_supplement_use_int</th>\n",
       "      <td>assessment_health_history_supplement_use_int</td>\n",
       "      <td>-0.024872</td>\n",
       "      <td>-0.447417</td>\n",
       "      <td>0.654676</td>\n",
       "      <td>957.0</td>\n",
       "      <td>0.859836</td>\n",
       "      <td>(0) No | (1) Rarely/Occasionally | (2) Yes | nan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                              feature  \\\n",
       "assessment_health_history_prescription_medicati...  assessment_health_history_prescription_medicat...   \n",
       "assessment_health_history_drugs_asthma_int                 assessment_health_history_drugs_asthma_int   \n",
       "assessment_health_history_drugs_allergy_int               assessment_health_history_drugs_allergy_int   \n",
       "assessment_health_history_drugs_anti_coagulatio...  assessment_health_history_drugs_anti_coagulati...   \n",
       "assessment_health_history_drugs_osteoporosis_int     assessment_health_history_drugs_osteoporosis_int   \n",
       "assessment_health_history_drugs_psychiatric_int       assessment_health_history_drugs_psychiatric_int   \n",
       "assessment_health_history_over_the_counter_medi...  assessment_health_history_over_the_counter_med...   \n",
       "assessment_health_history_supplement_use_int             assessment_health_history_supplement_use_int   \n",
       "\n",
       "                                                        beta  t_statistic  \\\n",
       "assessment_health_history_prescription_medicati... -0.087469    -2.585243   \n",
       "assessment_health_history_drugs_asthma_int         -0.080982    -1.932916   \n",
       "assessment_health_history_drugs_allergy_int        -0.061368    -1.214417   \n",
       "assessment_health_history_drugs_anti_coagulatio... -0.005262    -0.176631   \n",
       "assessment_health_history_drugs_osteoporosis_int   -0.004209    -0.215083   \n",
       "assessment_health_history_drugs_psychiatric_int    -0.028410    -0.657168   \n",
       "assessment_health_history_over_the_counter_medi...  0.024875     0.486631   \n",
       "assessment_health_history_supplement_use_int       -0.024872    -0.447417   \n",
       "\n",
       "                                                           p      n         q  \\\n",
       "assessment_health_history_prescription_medicati...  0.009878  964.0  0.079021   \n",
       "assessment_health_history_drugs_asthma_int          0.053542  961.0  0.214166   \n",
       "assessment_health_history_drugs_allergy_int         0.224889  961.0  0.599703   \n",
       "assessment_health_history_drugs_anti_coagulatio...  0.859836  960.0  0.859836   \n",
       "assessment_health_history_drugs_osteoporosis_int    0.829749  961.0  0.859836   \n",
       "assessment_health_history_drugs_psychiatric_int     0.511231  961.0  0.859836   \n",
       "assessment_health_history_over_the_counter_medi...  0.626632  957.0  0.859836   \n",
       "assessment_health_history_supplement_use_int        0.654676  957.0  0.859836   \n",
       "\n",
       "                                                                                              scales  \n",
       "assessment_health_history_prescription_medicati...                            (1) No | (2) Yes | nan  \n",
       "assessment_health_history_drugs_asthma_int            (0) Never | (1) Currently | (2) Formerly | nan  \n",
       "assessment_health_history_drugs_allergy_int           (0) Never | (1) Currently | (2) Formerly | nan  \n",
       "assessment_health_history_drugs_anti_coagulatio...    (0) Never | (1) Currently | (2) Formerly | nan  \n",
       "assessment_health_history_drugs_osteoporosis_int      (0) Never | (1) Currently | (2) Formerly | nan  \n",
       "assessment_health_history_drugs_psychiatric_int       (0) Never | (1) Currently | (2) Formerly | nan  \n",
       "assessment_health_history_over_the_counter_medi...  (0) No | (1) Rarely/Occasionally | (2) Yes | nan  \n",
       "assessment_health_history_supplement_use_int        (0) No | (1) Rarely/Occasionally | (2) Yes | nan  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tests_qs.sort_values(by=\"q\").head(10)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
