{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ae484767-d81e-4a22-956e-7a2dc63f9554",
   "metadata": {},
   "source": [
    "# Looking for associations with microbiome-derived metrics\n",
    "\n",
    "Okay, here is the pitch: suppose you have a novel metric/measure derived on microbiome data and you already calculated this measure for the amplicon sequencing samples in Arivale. What you want to do know is to find feature in the Arivale multiomics data that are associated/correlated with this new measure. This notebook will show you how to do that in a decently efficient manner.\n",
    "\n",
    "For the purpose of having an example we will use alpha diversity as an example. That's really not a novel measure but is one we already have. Let's start by getting the alpha diversity from the Arivale data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b1cac70c-2862-4326-b256-35d3ada1c429",
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'arivale_data_interface'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_7714/2561576713.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mwarnings\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0marivale_data_interface\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mget_snapshot\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mwarnings\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msimplefilter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"ignore\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'arivale_data_interface'"
     ]
    }
   ],
   "source": [
    "import warnings\n",
    "from arivale_data_interface import get_snapshot\n",
    "\n",
    "warnings.simplefilter(\"ignore\")\n",
    "\n",
    "div = get_snapshot(\"microbiome_diversity\")\n",
    "div.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42ccf5a5-7181-4b7e-a43e-7ed310d78fab",
   "metadata": {},
   "source": [
    "The minimum information you will need to keep around is the measure, the public_client_id, vendor_observation_id and days_in_program. If your data only has the vendor_observation_id than you can merge with the the columns in this table to get those other ones. So let's assemble our primary table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e6967d26-c9c5-46ad-b59e-c474d6e2808f",
   "metadata": {},
   "outputs": [],
   "source": [
    "measure = div[[\"diversity_shannon\", \"vendor_observation_id\", \"public_client_id\", \"days_in_program\", \"vendor_dashboard\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7b5b5e5-bb93-4d48-b129-bda481ce192f",
   "metadata": {},
   "source": [
    "Okay that is what we start with. Let's continue."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ec1d105-d5f0-40e9-aee9-4cdba01e885c",
   "metadata": {},
   "source": [
    "## Adding in the confounders\n",
    "\n",
    "The next thing is the confounders, e.g. covariates we want to correct for since they may be associated with our measure and the omics data without having a direct measure <-> omics data connection. For the microbiome data the principal ones are sex, age and BMI, but you could also correct for additional ones like genetic ancestry for instance. We find those measures in the `clients` and `weights` table and will add those in as well. For the weights we encounter a common challenge. There are multiple measurement points for each individual, so which one do we use for the BMI? We will use the closest data point to the microbiome sample. That is why we kept the `days_in_program` data around."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "517e7f28-a129-479b-8333-844288f630ae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5081, 8)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "clients = get_snapshot(\"clients\")[[\"public_client_id\", \"sex\", \"age\"]]\n",
    "measure = pd.merge(measure, clients, on=\"public_client_id\", how=\"inner\")\n",
    "\n",
    "weights = get_snapshot(\"weight\")[[\"BMI_CALC\", \"public_client_id\", \"days_in_program\"]].dropna()\n",
    "measure = pd.merge_asof(\n",
    "    measure[measure.public_client_id.isin(weights.public_client_id)].sort_values(by=\"days_in_program\"), \n",
    "    weights.sort_values(by=\"days_in_program\"), \n",
    "    by=\"public_client_id\", on=\"days_in_program\", direction=\"nearest\")\n",
    "\n",
    "measure.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c335f90-2560-434b-a0c2-4144475b28e6",
   "metadata": {},
   "source": [
    "Okay, some individuals were dropped due to missing data. No biggie. But at least we are now ready to actually start with our association analysis. We will start with the metabolomics data to illustrate the steps for the common omics analyses."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67ea0e16-ed10-4d07-b455-84c3e476970d",
   "metadata": {},
   "source": [
    "## Metabolome\n",
    "\n",
    "Okay each of those analysis will essentially use the same steps which are:\n",
    "1. identify metadata and actual data columns\n",
    "2. merge the omics data with the measure\n",
    "3. transform or normalize and filter features\n",
    "4. go through each feature and merge it with the metadata\n",
    "5. run the regression for each feature and track the results\n",
    "6. merge the results with the feature annotations\n",
    "\n",
    "Let's go then.\n",
    "\n",
    "### 1. Identify metadata and actual data columns\n",
    "\n",
    "We do this so we know which columns denote metadata and which ones actual measures (metabolite abundances)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a0b43a4a-ca19-41dd-8c3f-aae2951c6189",
   "metadata": {},
   "outputs": [],
   "source": [
    "metabolites = get_snapshot(\"metabolomics_corrected\").sort_values(by=\"days_in_program\")\n",
    "measure = measure.sort_values(by=\"days_in_program\")\n",
    "metabolite_features = metabolites.columns[8:]\n",
    "metabolites.days_in_program = metabolites.days_in_program.astype(\"float64\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3879e15-2b7c-44a9-b99b-b1b0a56eb64f",
   "metadata": {},
   "source": [
    "### 2. Merge metadata and measure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1a41a0cb-758f-44c1-be57-315f5b346351",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>diversity_shannon</th>\n",
       "      <th>vendor_observation_id</th>\n",
       "      <th>public_client_id</th>\n",
       "      <th>days_in_program</th>\n",
       "      <th>vendor_dashboard</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>BMI_CALC</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>days_since_first_call</th>\n",
       "      <th>...</th>\n",
       "      <th>999953114</th>\n",
       "      <th>999953157</th>\n",
       "      <th>999953172</th>\n",
       "      <th>999953266</th>\n",
       "      <th>999953267</th>\n",
       "      <th>999954831</th>\n",
       "      <th>999954832</th>\n",
       "      <th>999954834</th>\n",
       "      <th>999954839</th>\n",
       "      <th>999954840</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>4.542282</td>\n",
       "      <td>AV15-3636</td>\n",
       "      <td>01447446</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Second Genome</td>\n",
       "      <td>F</td>\n",
       "      <td>56.0</td>\n",
       "      <td>21.950816</td>\n",
       "      <td>A904AZ485-002</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.522402</td>\n",
       "      <td>1.073743</td>\n",
       "      <td>1.063270</td>\n",
       "      <td>1.146081</td>\n",
       "      <td>0.942299</td>\n",
       "      <td>1.696602</td>\n",
       "      <td>1.564450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>4.403993</td>\n",
       "      <td>AV15-1094</td>\n",
       "      <td>01789546</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Second Genome</td>\n",
       "      <td>F</td>\n",
       "      <td>87.0</td>\n",
       "      <td>18.692937</td>\n",
       "      <td>A976AR221-002</td>\n",
       "      <td>-6.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.153663</td>\n",
       "      <td>0.658271</td>\n",
       "      <td>0.338122</td>\n",
       "      <td>1.162175</td>\n",
       "      <td>1.845431</td>\n",
       "      <td>0.963192</td>\n",
       "      <td>0.942735</td>\n",
       "      <td>0.961368</td>\n",
       "      <td>0.650608</td>\n",
       "      <td>1.703106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3.318960</td>\n",
       "      <td>AV15-3588</td>\n",
       "      <td>01949838</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Second Genome</td>\n",
       "      <td>F</td>\n",
       "      <td>32.0</td>\n",
       "      <td>25.328546</td>\n",
       "      <td>A904AZ529-002</td>\n",
       "      <td>-3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.714835</td>\n",
       "      <td>1.276385</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.759459</td>\n",
       "      <td>1.205369</td>\n",
       "      <td>1.023168</td>\n",
       "      <td>1.106427</td>\n",
       "      <td>0.968127</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.118569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>4.283908</td>\n",
       "      <td>AV15-1295</td>\n",
       "      <td>01421882</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Second Genome</td>\n",
       "      <td>F</td>\n",
       "      <td>19.0</td>\n",
       "      <td>23.793846</td>\n",
       "      <td>A752AT612-002</td>\n",
       "      <td>-5.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.147181</td>\n",
       "      <td>0.758275</td>\n",
       "      <td>1.814588</td>\n",
       "      <td>3.996297</td>\n",
       "      <td>0.935515</td>\n",
       "      <td>0.915287</td>\n",
       "      <td>0.867081</td>\n",
       "      <td>0.871120</td>\n",
       "      <td>1.224561</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>4.278019</td>\n",
       "      <td>AV15-1348</td>\n",
       "      <td>01306924</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Second Genome</td>\n",
       "      <td>M</td>\n",
       "      <td>56.0</td>\n",
       "      <td>26.541837</td>\n",
       "      <td>A422AU094-002</td>\n",
       "      <td>-155.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.528782</td>\n",
       "      <td>1.606430</td>\n",
       "      <td>1.125981</td>\n",
       "      <td>0.896099</td>\n",
       "      <td>0.838319</td>\n",
       "      <td>1.023423</td>\n",
       "      <td>1.075749</td>\n",
       "      <td>1.051521</td>\n",
       "      <td>0.766951</td>\n",
       "      <td>1.026390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4468</th>\n",
       "      <td>3.684504</td>\n",
       "      <td>AV15-2674</td>\n",
       "      <td>01801678</td>\n",
       "      <td>453.0</td>\n",
       "      <td>Second Genome</td>\n",
       "      <td>M</td>\n",
       "      <td>35.0</td>\n",
       "      <td>20.705900</td>\n",
       "      <td>A706BA371-002</td>\n",
       "      <td>384.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.514679</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.768327</td>\n",
       "      <td>1.997009</td>\n",
       "      <td>1.036581</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.570969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4485</th>\n",
       "      <td>2.850451</td>\n",
       "      <td>AV15-2217</td>\n",
       "      <td>01874590</td>\n",
       "      <td>463.0</td>\n",
       "      <td>Second Genome</td>\n",
       "      <td>M</td>\n",
       "      <td>36.0</td>\n",
       "      <td>21.558261</td>\n",
       "      <td>A928BB321-002</td>\n",
       "      <td>384.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.382292</td>\n",
       "      <td>0.963556</td>\n",
       "      <td>1.770288</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.829983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4551</th>\n",
       "      <td>3.914202</td>\n",
       "      <td>22001805511808</td>\n",
       "      <td>01941857</td>\n",
       "      <td>509.0</td>\n",
       "      <td>research-microbiome</td>\n",
       "      <td>M</td>\n",
       "      <td>54.0</td>\n",
       "      <td>22.148708</td>\n",
       "      <td>A143BO102-003</td>\n",
       "      <td>488.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.016756</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.035061</td>\n",
       "      <td>0.566952</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.032233</td>\n",
       "      <td>1.101631</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4568</th>\n",
       "      <td>4.352964</td>\n",
       "      <td>22001612561459</td>\n",
       "      <td>01524770</td>\n",
       "      <td>517.0</td>\n",
       "      <td>research-microbiome</td>\n",
       "      <td>F</td>\n",
       "      <td>53.0</td>\n",
       "      <td>21.797160</td>\n",
       "      <td>A404BO077-008</td>\n",
       "      <td>531.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.649832</td>\n",
       "      <td>1.048798</td>\n",
       "      <td>3.931804</td>\n",
       "      <td>1.391538</td>\n",
       "      <td>1.358102</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.490642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4715</th>\n",
       "      <td>3.891328</td>\n",
       "      <td>22001805510258</td>\n",
       "      <td>01952180</td>\n",
       "      <td>696.0</td>\n",
       "      <td>research-microbiome</td>\n",
       "      <td>M</td>\n",
       "      <td>55.0</td>\n",
       "      <td>29.257965</td>\n",
       "      <td>A384BN840-008</td>\n",
       "      <td>671.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.733094</td>\n",
       "      <td>1.827141</td>\n",
       "      <td>0.985942</td>\n",
       "      <td>1.285743</td>\n",
       "      <td>0.628764</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.687353</td>\n",
       "      <td>1.597696</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1778 rows × 1310 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      diversity_shannon vendor_observation_id public_client_id  \\\n",
       "5              4.542282             AV15-3636         01447446   \n",
       "8              4.403993             AV15-1094         01789546   \n",
       "9              3.318960             AV15-3588         01949838   \n",
       "10             4.283908             AV15-1295         01421882   \n",
       "11             4.278019             AV15-1348         01306924   \n",
       "...                 ...                   ...              ...   \n",
       "4468           3.684504             AV15-2674         01801678   \n",
       "4485           2.850451             AV15-2217         01874590   \n",
       "4551           3.914202        22001805511808         01941857   \n",
       "4568           4.352964        22001612561459         01524770   \n",
       "4715           3.891328        22001805510258         01952180   \n",
       "\n",
       "      days_in_program     vendor_dashboard sex   age   BMI_CALC  \\\n",
       "5                 1.0        Second Genome   F  56.0  21.950816   \n",
       "8                 1.0        Second Genome   F  87.0  18.692937   \n",
       "9                 1.0        Second Genome   F  32.0  25.328546   \n",
       "10                1.0        Second Genome   F  19.0  23.793846   \n",
       "11                1.0        Second Genome   M  56.0  26.541837   \n",
       "...               ...                  ...  ..   ...        ...   \n",
       "4468            453.0        Second Genome   M  35.0  20.705900   \n",
       "4485            463.0        Second Genome   M  36.0  21.558261   \n",
       "4551            509.0  research-microbiome   M  54.0  22.148708   \n",
       "4568            517.0  research-microbiome   F  53.0  21.797160   \n",
       "4715            696.0  research-microbiome   M  55.0  29.257965   \n",
       "\n",
       "          sample_id  days_since_first_call  ...  999953114 999953157  \\\n",
       "5     A904AZ485-002                    1.0  ...        NaN       NaN   \n",
       "8     A976AR221-002                   -6.0  ...   1.153663  0.658271   \n",
       "9     A904AZ529-002                   -3.0  ...   0.714835  1.276385   \n",
       "10    A752AT612-002                   -5.0  ...   1.147181  0.758275   \n",
       "11    A422AU094-002                 -155.0  ...   1.528782  1.606430   \n",
       "...             ...                    ...  ...        ...       ...   \n",
       "4468  A706BA371-002                  384.0  ...   0.514679       NaN   \n",
       "4485  A928BB321-002                  384.0  ...        NaN       NaN   \n",
       "4551  A143BO102-003                  488.0  ...        NaN  1.016756   \n",
       "4568  A404BO077-008                  531.0  ...   1.649832  1.048798   \n",
       "4715  A384BN840-008                  671.0  ...   1.733094  1.827141   \n",
       "\n",
       "     999953172 999953266  999953267  999954831  999954832  999954834  \\\n",
       "5          NaN  1.522402   1.073743   1.063270   1.146081   0.942299   \n",
       "8     0.338122  1.162175   1.845431   0.963192   0.942735   0.961368   \n",
       "9          NaN  0.759459   1.205369   1.023168   1.106427   0.968127   \n",
       "10    1.814588  3.996297   0.935515   0.915287   0.867081   0.871120   \n",
       "11    1.125981  0.896099   0.838319   1.023423   1.075749   1.051521   \n",
       "...        ...       ...        ...        ...        ...        ...   \n",
       "4468  0.768327  1.997009   1.036581        NaN        NaN        NaN   \n",
       "4485  1.382292  0.963556   1.770288        NaN        NaN        NaN   \n",
       "4551       NaN  1.035061   0.566952        NaN        NaN        NaN   \n",
       "4568  3.931804  1.391538   1.358102        NaN        NaN        NaN   \n",
       "4715  0.985942  1.285743   0.628764        NaN        NaN        NaN   \n",
       "\n",
       "      999954839  999954840  \n",
       "5      1.696602   1.564450  \n",
       "8      0.650608   1.703106  \n",
       "9           NaN   1.118569  \n",
       "10     1.224561        NaN  \n",
       "11     0.766951   1.026390  \n",
       "...         ...        ...  \n",
       "4468        NaN   0.570969  \n",
       "4485        NaN   0.829983  \n",
       "4551   1.032233   1.101631  \n",
       "4568        NaN   0.490642  \n",
       "4715   0.687353   1.597696  \n",
       "\n",
       "[1778 rows x 1310 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metabolites_merged = pd.merge_asof(\n",
    "    measure, metabolites, \n",
    "    by=\"public_client_id\", \n",
    "    on=\"days_in_program\", \n",
    "    direction=\"nearest\", \n",
    "    tolerance=30.0).dropna(subset=metabolite_features, how=\"all\")\n",
    "metabolite_features = metabolites.columns[metabolites.columns.isin(metabolite_features)]\n",
    "metabolites_merged"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c024818d-1fc8-46af-aca3-bd2403db74fb",
   "metadata": {},
   "source": [
    "Okay we reduced the data set quite by a bit because we requested that we only want samples where blood and fecal samples were talken within 30 days. You can adjust this cutoof however you see fit. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1a3f0a4-e09d-4feb-937d-4ae6bf1d9b73",
   "metadata": {},
   "source": [
    "### 3. Preprocess data\n",
    "\n",
    "This depends a bit on your input but we usually wan to transform our features in a ways that they are comparable and not too rare. So here let's consider only metabolites that appear in at least 75% of all samples. The values are also all positive so we will log-transform and standardize them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2435d101-ff86-4707-b3ba-1c758223fa50",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1778, 929)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bad = metabolites_merged[metabolite_features].isnull().sum() / metabolites_merged.shape[0] > 0.25\n",
    "metabolite_features = bad[~bad].index\n",
    "metabolites_merged = metabolites_merged.drop(columns=bad[bad].index)\n",
    "metabolites_merged.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "557a8953-88b5-42dd-ba6d-eba712ea09f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "metabolites_merged[metabolite_features] = np.log(metabolites_merged[metabolite_features])\n",
    "metabolites_merged[metabolite_features] = metabolites_merged[metabolite_features].apply(lambda x: (x - x.mean()) / x.std())\n",
    "metabolites_merged.rename(columns=dict(zip(metabolite_features, \"metabolite_\" + metabolite_features)), inplace=True)\n",
    "metabolite_features = \"metabolite_\" + metabolite_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "aeb01e4a-da1c-4fd5-8ef7-fdcbf5741ffe",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert (metabolites_merged[metabolite_features].mean().abs() < 1e-6).all()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a685acd1-a698-40a2-8286-e0886c5a64d8",
   "metadata": {},
   "source": [
    "### 4. Run the regressions\n",
    "\n",
    "Now we have a clean data set and can start to run the regressions. You will see that this will surprisingly make use of all cores of the machine. This is because conda is setup by default to parallelize basic numeric operations like matrix factorization. Neat!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0941f2c5-b46f-4025-b114-50a82477321b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f4a784a64af044c9a3b9cd1e97291b2a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Output()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"></pre>\n"
      ],
      "text/plain": []
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from rich.progress import track\n",
    "from statsmodels.formula.api import ols\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "def get_results(feature):\n",
    "    \"\"\"Get a single association.\n",
    "    \n",
    "    As long as `args` and `df` are assigned this can be used on\n",
    "    any data set.\n",
    "    \"\"\"\n",
    "    formula = f\"{feature} ~ C(sex) + age + BMI_CALC + C(vendor_dashboard) + diversity_shannon\"\n",
    "    fitted = ols(formula, data=df).fit()\n",
    "    return pd.DataFrame({\n",
    "        \"feature\": feature,\n",
    "        \"beta\": fitted.params[\"diversity_shannon\"],\n",
    "        \"t_statistic\": fitted.tvalues[\"diversity_shannon\"],\n",
    "        \"p\": fitted.pvalues[\"diversity_shannon\"],\n",
    "        \"n\": fitted.nobs\n",
    "        }, index=[feature])\n",
    "\n",
    "args = metabolite_features\n",
    "df = metabolites_merged\n",
    "results = map(get_results, track(args))\n",
    "results = list(results)\n",
    "tests = pd.concat(results)\n",
    "tests[\"q\"] = multipletests(tests.p, method=\"fdr_bh\")[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb5a0b05-f4f7-462e-ab4b-6eaa3946dd90",
   "metadata": {},
   "source": [
    "### 5. Merge in annotations\n",
    "\n",
    "Finally we merge in the metadata for the metabolites."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f8159e3c-f492-44a3-a34b-a125f649dac2",
   "metadata": {},
   "outputs": [],
   "source": [
    "anns = get_snapshot(\"metabolomics_metadata\").iloc[:, 0:8]\n",
    "anns[\"feature\"] = \"metabolite_\" + anns.CHEMICAL_ID.astype(str)\n",
    "tests = pd.merge(tests, anns, on=\"feature\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2b7fa05a-f54a-47d6-8f53-d0d0a320d118",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>beta</th>\n",
       "      <th>t_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>n</th>\n",
       "      <th>q</th>\n",
       "      <th>CHEMICAL_ID</th>\n",
       "      <th>SUB_PATHWAY</th>\n",
       "      <th>SUPER_PATHWAY</th>\n",
       "      <th>BIOCHEMICAL_NAME</th>\n",
       "      <th>CAS</th>\n",
       "      <th>KEGG</th>\n",
       "      <th>HMDB</th>\n",
       "      <th>PUBCHEM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>420</th>\n",
       "      <td>metabolite_100002253</td>\n",
       "      <td>1.078235</td>\n",
       "      <td>20.705369</td>\n",
       "      <td>1.942997e-84</td>\n",
       "      <td>1583.0</td>\n",
       "      <td>1.777843e-81</td>\n",
       "      <td>100002253</td>\n",
       "      <td>Food Component/Plant</td>\n",
       "      <td>Xenobiotics</td>\n",
       "      <td>cinnamoylglycine</td>\n",
       "      <td>16534-24-0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HMDB11621</td>\n",
       "      <td>709625.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>metabolite_100000010</td>\n",
       "      <td>1.052296</td>\n",
       "      <td>19.073275</td>\n",
       "      <td>5.456473e-73</td>\n",
       "      <td>1549.0</td>\n",
       "      <td>2.496336e-70</td>\n",
       "      <td>100000010</td>\n",
       "      <td>Phenylalanine and Tyrosine Metabolism</td>\n",
       "      <td>Amino Acid</td>\n",
       "      <td>3-phenylpropionate (hydrocinnamate)</td>\n",
       "      <td>501-52-0</td>\n",
       "      <td>C05629</td>\n",
       "      <td>HMDB00764</td>\n",
       "      <td>107.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>828</th>\n",
       "      <td>metabolite_999946613</td>\n",
       "      <td>0.975777</td>\n",
       "      <td>18.860249</td>\n",
       "      <td>5.209847e-72</td>\n",
       "      <td>1657.0</td>\n",
       "      <td>1.191752e-69</td>\n",
       "      <td>999946613</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X - 12216</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>425</th>\n",
       "      <td>metabolite_100002488</td>\n",
       "      <td>-0.976546</td>\n",
       "      <td>-18.867847</td>\n",
       "      <td>4.465597e-72</td>\n",
       "      <td>1661.0</td>\n",
       "      <td>1.191752e-69</td>\n",
       "      <td>100002488</td>\n",
       "      <td>Secondary Bile Acid Metabolism</td>\n",
       "      <td>Lipid</td>\n",
       "      <td>isoursodeoxycholate</td>\n",
       "      <td>78919-26-3</td>\n",
       "      <td>C17662</td>\n",
       "      <td>HMDB00686</td>\n",
       "      <td>127601.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>metabolite_100001315</td>\n",
       "      <td>0.817917</td>\n",
       "      <td>15.511253</td>\n",
       "      <td>7.746503e-51</td>\n",
       "      <td>1698.0</td>\n",
       "      <td>1.417610e-48</td>\n",
       "      <td>100001315</td>\n",
       "      <td>Phenylalanine and Tyrosine Metabolism</td>\n",
       "      <td>Amino Acid</td>\n",
       "      <td>p-cresol sulfate</td>\n",
       "      <td>3233-57-7</td>\n",
       "      <td>C01468</td>\n",
       "      <td>HMDB11635</td>\n",
       "      <td>4615423.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>metabolite_100000014</td>\n",
       "      <td>0.793217</td>\n",
       "      <td>15.195796</td>\n",
       "      <td>6.218955e-49</td>\n",
       "      <td>1669.0</td>\n",
       "      <td>9.483907e-47</td>\n",
       "      <td>100000014</td>\n",
       "      <td>Benzoate Metabolism</td>\n",
       "      <td>Xenobiotics</td>\n",
       "      <td>hippurate</td>\n",
       "      <td>495-69-2</td>\n",
       "      <td>C01586</td>\n",
       "      <td>HMDB00714</td>\n",
       "      <td>464.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395</th>\n",
       "      <td>metabolite_100002021</td>\n",
       "      <td>0.794902</td>\n",
       "      <td>14.649483</td>\n",
       "      <td>2.093177e-45</td>\n",
       "      <td>1448.0</td>\n",
       "      <td>2.736081e-43</td>\n",
       "      <td>100002021</td>\n",
       "      <td>Steroid</td>\n",
       "      <td>Lipid</td>\n",
       "      <td>5alpha-androstan-3beta,17alpha-diol disulfate</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>812</th>\n",
       "      <td>metabolite_999946507</td>\n",
       "      <td>0.883826</td>\n",
       "      <td>13.722354</td>\n",
       "      <td>2.555802e-40</td>\n",
       "      <td>1417.0</td>\n",
       "      <td>2.923198e-38</td>\n",
       "      <td>999946507</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X - 11850</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>metabolite_100001657</td>\n",
       "      <td>0.703288</td>\n",
       "      <td>12.838754</td>\n",
       "      <td>4.714016e-36</td>\n",
       "      <td>1679.0</td>\n",
       "      <td>4.792583e-34</td>\n",
       "      <td>100001657</td>\n",
       "      <td>Secondary Bile Acid Metabolism</td>\n",
       "      <td>Lipid</td>\n",
       "      <td>glycolithocholate sulfate*</td>\n",
       "      <td>15324-64-8</td>\n",
       "      <td>C11301</td>\n",
       "      <td>HMDB02639</td>\n",
       "      <td>72222.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>438</th>\n",
       "      <td>metabolite_100002911</td>\n",
       "      <td>-0.696692</td>\n",
       "      <td>-12.507278</td>\n",
       "      <td>2.287488e-34</td>\n",
       "      <td>1666.0</td>\n",
       "      <td>2.093051e-32</td>\n",
       "      <td>100002911</td>\n",
       "      <td>Secondary Bile Acid Metabolism</td>\n",
       "      <td>Lipid</td>\n",
       "      <td>glycoursodeoxycholate</td>\n",
       "      <td>64480-66-6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HMDB00708</td>\n",
       "      <td>12310288.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  feature      beta  t_statistic             p       n  \\\n",
       "420  metabolite_100002253  1.078235    20.705369  1.942997e-84  1583.0   \n",
       "167  metabolite_100000010  1.052296    19.073275  5.456473e-73  1549.0   \n",
       "828  metabolite_999946613  0.975777    18.860249  5.209847e-72  1657.0   \n",
       "425  metabolite_100002488 -0.976546   -18.867847  4.465597e-72  1661.0   \n",
       "283  metabolite_100001315  0.817917    15.511253  7.746503e-51  1698.0   \n",
       "168  metabolite_100000014  0.793217    15.195796  6.218955e-49  1669.0   \n",
       "395  metabolite_100002021  0.794902    14.649483  2.093177e-45  1448.0   \n",
       "812  metabolite_999946507  0.883826    13.722354  2.555802e-40  1417.0   \n",
       "355  metabolite_100001657  0.703288    12.838754  4.714016e-36  1679.0   \n",
       "438  metabolite_100002911 -0.696692   -12.507278  2.287488e-34  1666.0   \n",
       "\n",
       "                q  CHEMICAL_ID                            SUB_PATHWAY  \\\n",
       "420  1.777843e-81    100002253                   Food Component/Plant   \n",
       "167  2.496336e-70    100000010  Phenylalanine and Tyrosine Metabolism   \n",
       "828  1.191752e-69    999946613                                    NaN   \n",
       "425  1.191752e-69    100002488         Secondary Bile Acid Metabolism   \n",
       "283  1.417610e-48    100001315  Phenylalanine and Tyrosine Metabolism   \n",
       "168  9.483907e-47    100000014                    Benzoate Metabolism   \n",
       "395  2.736081e-43    100002021                                Steroid   \n",
       "812  2.923198e-38    999946507                                    NaN   \n",
       "355  4.792583e-34    100001657         Secondary Bile Acid Metabolism   \n",
       "438  2.093051e-32    100002911         Secondary Bile Acid Metabolism   \n",
       "\n",
       "    SUPER_PATHWAY                               BIOCHEMICAL_NAME         CAS  \\\n",
       "420   Xenobiotics                               cinnamoylglycine  16534-24-0   \n",
       "167    Amino Acid            3-phenylpropionate (hydrocinnamate)    501-52-0   \n",
       "828           NaN                                      X - 12216         NaN   \n",
       "425         Lipid                            isoursodeoxycholate  78919-26-3   \n",
       "283    Amino Acid                               p-cresol sulfate   3233-57-7   \n",
       "168   Xenobiotics                                      hippurate    495-69-2   \n",
       "395         Lipid  5alpha-androstan-3beta,17alpha-diol disulfate         NaN   \n",
       "812           NaN                                      X - 11850         NaN   \n",
       "355         Lipid                     glycolithocholate sulfate*  15324-64-8   \n",
       "438         Lipid                          glycoursodeoxycholate  64480-66-6   \n",
       "\n",
       "       KEGG       HMDB     PUBCHEM  \n",
       "420     NaN  HMDB11621    709625.0  \n",
       "167  C05629  HMDB00764       107.0  \n",
       "828     NaN        NaN         NaN  \n",
       "425  C17662  HMDB00686    127601.0  \n",
       "283  C01468  HMDB11635   4615423.0  \n",
       "168  C01586  HMDB00714       464.0  \n",
       "395     NaN        NaN         NaN  \n",
       "812     NaN        NaN         NaN  \n",
       "355  C11301  HMDB02639     72222.0  \n",
       "438     NaN  HMDB00708  12310288.0  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# tests.to_csv(\"metabolome_tests.csv\", index=False)  # uncomment to save the result\n",
    "tests.sort_values(by=\"q\").head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5a40826-af01-4fda-a25c-8eea1a1f503c",
   "metadata": {},
   "source": [
    "Okay, that was this. You could do exactly the same for proteomics. Let's run through it with clinical labs here since we don't really want to adjust the variables that much. We will group the steps this time.\n",
    "\n",
    "## Clinical labs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5d3f8b7a-aa3c-456e-ab66-21ac754a541a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find the features\n",
    "chems = get_snapshot(\"chemistries\", clean=True).sort_values(by=\"days_in_program\")\n",
    "chem_features = chems.columns[12:]  # inspect the table to find out where the metadata ends \n",
    "chems.days_in_program = chems.days_in_program.astype(\"float64\")\n",
    "\n",
    "# Merge with measure\n",
    "chems_merged = pd.merge_asof(\n",
    "    measure, chems, \n",
    "    by=\"public_client_id\", \n",
    "    on=\"days_in_program\", \n",
    "    direction=\"nearest\", \n",
    "    tolerance=30.0).dropna(subset=chem_features, how=\"all\")\n",
    "chem_features = chems.columns[chems.columns.isin(chem_features)]\n",
    "\n",
    "# Filter uncomplete\n",
    "bad = chems_merged[chem_features].isnull().sum() / chems_merged.shape[0] > 0.25\n",
    "chem_features = bad[~bad].index\n",
    "chems_merged = chems_merged.drop(columns=bad[bad].index)\n",
    "\n",
    "# We don't transform the values further since some of them are integer etc. and they have associated units we don't want to break\n",
    "# However, that mean we can't compare the beta values between features."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f14ee2ec-468e-4d2f-a9a7-569585abee78",
   "metadata": {},
   "source": [
    "Looks pretty concise this way and we can start to run the testing loop again."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "4969e0e2-6d56-48d6-afff-a2bca028cdf4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "eab7cf4843bd4f6faa8b70eba56f411b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Output()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"></pre>\n"
      ],
      "text/plain": []
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "args = chem_features\n",
    "df = chems_merged  # don't forget to reassign those\n",
    "results = map(get_results, track(args))\n",
    "results = list(results)\n",
    "tests_chems = pd.concat(results)\n",
    "tests_chems[\"q\"] = multipletests(tests_chems.p, method=\"fdr_bh\")[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "7ddb4d9b-c636-41ef-a326-4162236f54bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "anns = get_snapshot(\"chemistries_metadata\")\n",
    "anns[\"feature\"] = anns.Name\n",
    "tests_chems = pd.merge(tests_chems, anns, on=\"feature\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "794d1c1f-457d-42a0-a11e-70c145fc5cda",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>beta</th>\n",
       "      <th>t_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>n</th>\n",
       "      <th>q</th>\n",
       "      <th>Name</th>\n",
       "      <th>Display Name</th>\n",
       "      <th>Labcorp ID</th>\n",
       "      <th>Labcorp Name</th>\n",
       "      <th>Labcorp LOINC ID</th>\n",
       "      <th>Labcorp LOINC Name</th>\n",
       "      <th>Quest ID</th>\n",
       "      <th>Quest Name</th>\n",
       "      <th>Quest LOINC ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>TRIGLYCERIDES</td>\n",
       "      <td>-18.822177</td>\n",
       "      <td>-9.858740</td>\n",
       "      <td>1.136545e-22</td>\n",
       "      <td>4008.0</td>\n",
       "      <td>7.955817e-21</td>\n",
       "      <td>TRIGLYCERIDES</td>\n",
       "      <td>Triglycerides</td>\n",
       "      <td>884256.0</td>\n",
       "      <td>Triglycerides</td>\n",
       "      <td>2571-8</td>\n",
       "      <td>Triglyceride</td>\n",
       "      <td>XRI</td>\n",
       "      <td>TRIGLYCERIDES</td>\n",
       "      <td>2571-8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>LPIR_SCORE</td>\n",
       "      <td>-4.954704</td>\n",
       "      <td>-8.190493</td>\n",
       "      <td>3.605946e-16</td>\n",
       "      <td>3510.0</td>\n",
       "      <td>8.413874e-15</td>\n",
       "      <td>LPIR_SCORE</td>\n",
       "      <td>LP-IR Score</td>\n",
       "      <td>884312.0</td>\n",
       "      <td>LP-IR Score</td>\n",
       "      <td>62255-5</td>\n",
       "      <td>Lipoprotein insulin resistance score</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>LDL_SIZE</td>\n",
       "      <td>0.144971</td>\n",
       "      <td>7.612450</td>\n",
       "      <td>3.441094e-14</td>\n",
       "      <td>3491.0</td>\n",
       "      <td>4.817531e-13</td>\n",
       "      <td>LDL_SIZE</td>\n",
       "      <td>LDL Size</td>\n",
       "      <td>884309.0</td>\n",
       "      <td>LDL Size</td>\n",
       "      <td>17782-4</td>\n",
       "      <td>Lipoprotein.beta.subparticle</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GLOBULIN</td>\n",
       "      <td>-0.071668</td>\n",
       "      <td>-5.933660</td>\n",
       "      <td>3.213327e-09</td>\n",
       "      <td>4011.0</td>\n",
       "      <td>2.811661e-08</td>\n",
       "      <td>GLOBULIN</td>\n",
       "      <td>Globulin</td>\n",
       "      <td>12039.0</td>\n",
       "      <td>Globulin, Total</td>\n",
       "      <td>10834-0</td>\n",
       "      <td>Globulin</td>\n",
       "      <td>XGL</td>\n",
       "      <td>GLOBULIN</td>\n",
       "      <td>10834-0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>GLUCOSE</td>\n",
       "      <td>-3.110157</td>\n",
       "      <td>-5.090439</td>\n",
       "      <td>3.736450e-07</td>\n",
       "      <td>4011.0</td>\n",
       "      <td>2.377741e-06</td>\n",
       "      <td>GLUCOSE</td>\n",
       "      <td>Glucose</td>\n",
       "      <td>1032.0</td>\n",
       "      <td>Glucose</td>\n",
       "      <td>2345-7</td>\n",
       "      <td>Glucose</td>\n",
       "      <td>XGU</td>\n",
       "      <td>GLUCOSE</td>\n",
       "      <td>2345-7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>INSULIN</td>\n",
       "      <td>-1.533289</td>\n",
       "      <td>-5.030844</td>\n",
       "      <td>5.097897e-07</td>\n",
       "      <td>4002.0</td>\n",
       "      <td>2.973773e-06</td>\n",
       "      <td>INSULIN</td>\n",
       "      <td>Insulin</td>\n",
       "      <td>4333.0</td>\n",
       "      <td>Insulin</td>\n",
       "      <td>20448-7</td>\n",
       "      <td>Insulin</td>\n",
       "      <td>ISLN</td>\n",
       "      <td>INSULIN</td>\n",
       "      <td>20448-7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>OMEGA_6_TOTAL</td>\n",
       "      <td>0.515363</td>\n",
       "      <td>4.564454</td>\n",
       "      <td>5.179511e-06</td>\n",
       "      <td>3518.0</td>\n",
       "      <td>2.266036e-05</td>\n",
       "      <td>OMEGA_6_TOTAL</td>\n",
       "      <td>Omega-6 total</td>\n",
       "      <td>822904.0</td>\n",
       "      <td>Omega-6 total</td>\n",
       "      <td>SOLOINC</td>\n",
       "      <td>SENDOUT NO LOINC PROVIDED</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DHA</td>\n",
       "      <td>0.141907</td>\n",
       "      <td>4.574212</td>\n",
       "      <td>4.926423e-06</td>\n",
       "      <td>3965.0</td>\n",
       "      <td>2.266036e-05</td>\n",
       "      <td>DHA</td>\n",
       "      <td>DHA</td>\n",
       "      <td>817656.0</td>\n",
       "      <td>DHA</td>\n",
       "      <td>SOLOINC</td>\n",
       "      <td>SENDOUT NO LOINC PROVIDED</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GGT</td>\n",
       "      <td>-2.346285</td>\n",
       "      <td>-3.068895</td>\n",
       "      <td>2.163243e-03</td>\n",
       "      <td>3918.0</td>\n",
       "      <td>5.408108e-03</td>\n",
       "      <td>GGT</td>\n",
       "      <td>Gamma-glutamyl transpeptidase</td>\n",
       "      <td>1958.0</td>\n",
       "      <td>GGT</td>\n",
       "      <td>2324-2</td>\n",
       "      <td>Gamma glutamyl transferase</td>\n",
       "      <td>XGT</td>\n",
       "      <td>GGT</td>\n",
       "      <td>2324-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>FERRITIN</td>\n",
       "      <td>-12.120816</td>\n",
       "      <td>-2.761082</td>\n",
       "      <td>5.789299e-03</td>\n",
       "      <td>3724.0</td>\n",
       "      <td>1.307261e-02</td>\n",
       "      <td>FERRITIN</td>\n",
       "      <td>Ferritin</td>\n",
       "      <td>4598.0</td>\n",
       "      <td>Ferritin, Serum</td>\n",
       "      <td>2276-4</td>\n",
       "      <td>Ferritin</td>\n",
       "      <td>FRRN</td>\n",
       "      <td>FERRITIN</td>\n",
       "      <td>2276-4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          feature       beta  t_statistic             p       n             q  \\\n",
       "28  TRIGLYCERIDES -18.822177    -9.858740  1.136545e-22  4008.0  7.955817e-21   \n",
       "17     LPIR_SCORE  -4.954704    -8.190493  3.605946e-16  3510.0  8.413874e-15   \n",
       "15       LDL_SIZE   0.144971     7.612450  3.441094e-14  3491.0  4.817531e-13   \n",
       "10       GLOBULIN  -0.071668    -5.933660  3.213327e-09  4011.0  2.811661e-08   \n",
       "11        GLUCOSE  -3.110157    -5.090439  3.736450e-07  4011.0  2.377741e-06   \n",
       "14        INSULIN  -1.533289    -5.030844  5.097897e-07  4002.0  2.973773e-06   \n",
       "24  OMEGA_6_TOTAL   0.515363     4.564454  5.179511e-06  3518.0  2.266036e-05   \n",
       "4             DHA   0.141907     4.574212  4.926423e-06  3965.0  2.266036e-05   \n",
       "9             GGT  -2.346285    -3.068895  2.163243e-03  3918.0  5.408108e-03   \n",
       "8        FERRITIN -12.120816    -2.761082  5.789299e-03  3724.0  1.307261e-02   \n",
       "\n",
       "             Name                   Display Name  Labcorp ID     Labcorp Name  \\\n",
       "28  TRIGLYCERIDES                  Triglycerides    884256.0    Triglycerides   \n",
       "17     LPIR_SCORE                    LP-IR Score    884312.0      LP-IR Score   \n",
       "15       LDL_SIZE                       LDL Size    884309.0         LDL Size   \n",
       "10       GLOBULIN                       Globulin     12039.0  Globulin, Total   \n",
       "11        GLUCOSE                        Glucose      1032.0          Glucose   \n",
       "14        INSULIN                        Insulin      4333.0          Insulin   \n",
       "24  OMEGA_6_TOTAL                  Omega-6 total    822904.0    Omega-6 total   \n",
       "4             DHA                            DHA    817656.0              DHA   \n",
       "9             GGT  Gamma-glutamyl transpeptidase      1958.0              GGT   \n",
       "8        FERRITIN                       Ferritin      4598.0  Ferritin, Serum   \n",
       "\n",
       "   Labcorp LOINC ID                    Labcorp LOINC Name Quest ID  \\\n",
       "28           2571-8                          Triglyceride      XRI   \n",
       "17          62255-5  Lipoprotein insulin resistance score      NaN   \n",
       "15          17782-4          Lipoprotein.beta.subparticle      NaN   \n",
       "10          10834-0                              Globulin      XGL   \n",
       "11           2345-7                               Glucose      XGU   \n",
       "14          20448-7                               Insulin     ISLN   \n",
       "24          SOLOINC             SENDOUT NO LOINC PROVIDED      NaN   \n",
       "4           SOLOINC             SENDOUT NO LOINC PROVIDED      NaN   \n",
       "9            2324-2            Gamma glutamyl transferase      XGT   \n",
       "8            2276-4                              Ferritin     FRRN   \n",
       "\n",
       "       Quest Name Quest LOINC ID  \n",
       "28  TRIGLYCERIDES         2571-8  \n",
       "17            NaN            NaN  \n",
       "15            NaN            NaN  \n",
       "10       GLOBULIN        10834-0  \n",
       "11        GLUCOSE         2345-7  \n",
       "14        INSULIN        20448-7  \n",
       "24            NaN            NaN  \n",
       "4             NaN            NaN  \n",
       "9             GGT         2324-2  \n",
       "8        FERRITIN         2276-4  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tests_chems.sort_values(by=\"q\").head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd82876f-afa7-442f-8f17-2215966c8917",
   "metadata": {},
   "source": [
    "## Questionnaires\n",
    "\n",
    "Okay same things as before but we get the questionnaire data from the general assessments and will express it on an integer scale from 0-M where 0 is little and M is a lot. For that we will only use the \"enum\" and \"int\" scales."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "53047222-2fc8-4c71-8c07-6a506186fca9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>diversity_shannon</th>\n",
       "      <th>vendor_observation_id</th>\n",
       "      <th>public_client_id</th>\n",
       "      <th>days_in_program</th>\n",
       "      <th>vendor_dashboard</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>BMI_CALC</th>\n",
       "      <th>vendor</th>\n",
       "      <th>days_since_first_call</th>\n",
       "      <th>...</th>\n",
       "      <th>assessment_productivity_recent_ailments_anxiety</th>\n",
       "      <th>assessment_productivity_recent_ailments_depression</th>\n",
       "      <th>assessment_productivity_hours_missed_from_illness_real</th>\n",
       "      <th>assessment_productivity_hours_worked_real</th>\n",
       "      <th>assessment_lifestyle_avg_stress_enum</th>\n",
       "      <th>assessment_lifestyle_stress_management_enum</th>\n",
       "      <th>assessment_pss_four_item_going_your_way_enum</th>\n",
       "      <th>assessment_pss_four_item_handle_problems_enum</th>\n",
       "      <th>assessment_pss_four_item_insurmountable_enum</th>\n",
       "      <th>assessment_pss_four_item_out_of_control_enum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3.870313</td>\n",
       "      <td>AV15-4372</td>\n",
       "      <td>01602320</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>Second Genome</td>\n",
       "      <td>M</td>\n",
       "      <td>48.0</td>\n",
       "      <td>37.229730</td>\n",
       "      <td>Assessments</td>\n",
       "      <td>-5.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4.281822</td>\n",
       "      <td>AV15-3716</td>\n",
       "      <td>01228476</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Second Genome</td>\n",
       "      <td>F</td>\n",
       "      <td>51.0</td>\n",
       "      <td>22.462722</td>\n",
       "      <td>Assessments</td>\n",
       "      <td>-26.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3.818317</td>\n",
       "      <td>AV15-3627</td>\n",
       "      <td>01048590</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Second Genome</td>\n",
       "      <td>F</td>\n",
       "      <td>48.0</td>\n",
       "      <td>22.459158</td>\n",
       "      <td>Assessments</td>\n",
       "      <td>-4.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2.999081</td>\n",
       "      <td>AV15-3759</td>\n",
       "      <td>01250245</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Second Genome</td>\n",
       "      <td>M</td>\n",
       "      <td>32.0</td>\n",
       "      <td>46.955261</td>\n",
       "      <td>Assessments</td>\n",
       "      <td>-4.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2.496614</td>\n",
       "      <td>AV15-3489</td>\n",
       "      <td>01826852</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Second Genome</td>\n",
       "      <td>F</td>\n",
       "      <td>35.0</td>\n",
       "      <td>24.541709</td>\n",
       "      <td>Assessments</td>\n",
       "      <td>-27.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5074</th>\n",
       "      <td>3.354139</td>\n",
       "      <td>22001805510143</td>\n",
       "      <td>01108982</td>\n",
       "      <td>1354.0</td>\n",
       "      <td>research-microbiome</td>\n",
       "      <td>M</td>\n",
       "      <td>62.0</td>\n",
       "      <td>22.081310</td>\n",
       "      <td>Assessments</td>\n",
       "      <td>1311.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5076</th>\n",
       "      <td>3.868454</td>\n",
       "      <td>22001805521073</td>\n",
       "      <td>01441946</td>\n",
       "      <td>1367.0</td>\n",
       "      <td>research-microbiome</td>\n",
       "      <td>F</td>\n",
       "      <td>44.0</td>\n",
       "      <td>19.732234</td>\n",
       "      <td>Assessments</td>\n",
       "      <td>1334.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5077</th>\n",
       "      <td>4.627451</td>\n",
       "      <td>22001805520179</td>\n",
       "      <td>01152098</td>\n",
       "      <td>1368.0</td>\n",
       "      <td>research-microbiome</td>\n",
       "      <td>F</td>\n",
       "      <td>54.0</td>\n",
       "      <td>21.077601</td>\n",
       "      <td>Assessments</td>\n",
       "      <td>1351.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5078</th>\n",
       "      <td>4.207403</td>\n",
       "      <td>22001805510387</td>\n",
       "      <td>01023946</td>\n",
       "      <td>1371.0</td>\n",
       "      <td>research-microbiome</td>\n",
       "      <td>F</td>\n",
       "      <td>43.0</td>\n",
       "      <td>27.259184</td>\n",
       "      <td>Assessments</td>\n",
       "      <td>1339.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5079</th>\n",
       "      <td>4.273946</td>\n",
       "      <td>22001805521617</td>\n",
       "      <td>01483778</td>\n",
       "      <td>1377.0</td>\n",
       "      <td>research-microbiome</td>\n",
       "      <td>F</td>\n",
       "      <td>39.0</td>\n",
       "      <td>23.440253</td>\n",
       "      <td>Assessments</td>\n",
       "      <td>1324.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3616 rows × 3291 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      diversity_shannon vendor_observation_id public_client_id  \\\n",
       "0              3.870313             AV15-4372         01602320   \n",
       "1              4.281822             AV15-3716         01228476   \n",
       "2              3.818317             AV15-3627         01048590   \n",
       "3              2.999081             AV15-3759         01250245   \n",
       "4              2.496614             AV15-3489         01826852   \n",
       "...                 ...                   ...              ...   \n",
       "5074           3.354139        22001805510143         01108982   \n",
       "5076           3.868454        22001805521073         01441946   \n",
       "5077           4.627451        22001805520179         01152098   \n",
       "5078           4.207403        22001805510387         01023946   \n",
       "5079           4.273946        22001805521617         01483778   \n",
       "\n",
       "      days_in_program     vendor_dashboard sex   age   BMI_CALC       vendor  \\\n",
       "0                -2.0        Second Genome   M  48.0  37.229730  Assessments   \n",
       "1                 0.0        Second Genome   F  51.0  22.462722  Assessments   \n",
       "2                 1.0        Second Genome   F  48.0  22.459158  Assessments   \n",
       "3                 1.0        Second Genome   M  32.0  46.955261  Assessments   \n",
       "4                 1.0        Second Genome   F  35.0  24.541709  Assessments   \n",
       "...               ...                  ...  ..   ...        ...          ...   \n",
       "5074           1354.0  research-microbiome   M  62.0  22.081310  Assessments   \n",
       "5076           1367.0  research-microbiome   F  44.0  19.732234  Assessments   \n",
       "5077           1368.0  research-microbiome   F  54.0  21.077601  Assessments   \n",
       "5078           1371.0  research-microbiome   F  43.0  27.259184  Assessments   \n",
       "5079           1377.0  research-microbiome   F  39.0  23.440253  Assessments   \n",
       "\n",
       "      days_since_first_call  ...  \\\n",
       "0                      -5.0  ...   \n",
       "1                     -26.0  ...   \n",
       "2                      -4.0  ...   \n",
       "3                      -4.0  ...   \n",
       "4                     -27.0  ...   \n",
       "...                     ...  ...   \n",
       "5074                 1311.0  ...   \n",
       "5076                 1334.0  ...   \n",
       "5077                 1351.0  ...   \n",
       "5078                 1339.0  ...   \n",
       "5079                 1324.0  ...   \n",
       "\n",
       "      assessment_productivity_recent_ailments_anxiety  \\\n",
       "0                                                 NaN   \n",
       "1                                                 NaN   \n",
       "2                                                 NaN   \n",
       "3                                                 NaN   \n",
       "4                                                 NaN   \n",
       "...                                               ...   \n",
       "5074                                              NaN   \n",
       "5076                                              NaN   \n",
       "5077                                              NaN   \n",
       "5078                                              NaN   \n",
       "5079                                              NaN   \n",
       "\n",
       "     assessment_productivity_recent_ailments_depression  \\\n",
       "0                                                   NaN   \n",
       "1                                                   NaN   \n",
       "2                                                   NaN   \n",
       "3                                                   NaN   \n",
       "4                                                   NaN   \n",
       "...                                                 ...   \n",
       "5074                                                NaN   \n",
       "5076                                                NaN   \n",
       "5077                                                NaN   \n",
       "5078                                                NaN   \n",
       "5079                                                NaN   \n",
       "\n",
       "     assessment_productivity_hours_missed_from_illness_real  \\\n",
       "0                                                   NaN       \n",
       "1                                                   NaN       \n",
       "2                                                   NaN       \n",
       "3                                                   NaN       \n",
       "4                                                   NaN       \n",
       "...                                                 ...       \n",
       "5074                                                NaN       \n",
       "5076                                                NaN       \n",
       "5077                                                NaN       \n",
       "5078                                                NaN       \n",
       "5079                                                NaN       \n",
       "\n",
       "     assessment_productivity_hours_worked_real  \\\n",
       "0                                          NaN   \n",
       "1                                          NaN   \n",
       "2                                          NaN   \n",
       "3                                          NaN   \n",
       "4                                          NaN   \n",
       "...                                        ...   \n",
       "5074                                       NaN   \n",
       "5076                                       NaN   \n",
       "5077                                       NaN   \n",
       "5078                                       NaN   \n",
       "5079                                       NaN   \n",
       "\n",
       "      assessment_lifestyle_avg_stress_enum  \\\n",
       "0                                      7.0   \n",
       "1                                      4.0   \n",
       "2                                      0.0   \n",
       "3                                      7.0   \n",
       "4                                      6.0   \n",
       "...                                    ...   \n",
       "5074                                   1.0   \n",
       "5076                                   5.0   \n",
       "5077                                   1.0   \n",
       "5078                                   7.0   \n",
       "5079                                   5.0   \n",
       "\n",
       "      assessment_lifestyle_stress_management_enum  \\\n",
       "0                                             0.0   \n",
       "1                                             0.0   \n",
       "2                                             0.0   \n",
       "3                                             0.0   \n",
       "4                                             0.0   \n",
       "...                                           ...   \n",
       "5074                                          0.0   \n",
       "5076                                          1.0   \n",
       "5077                                          1.0   \n",
       "5078                                          1.0   \n",
       "5079                                          1.0   \n",
       "\n",
       "      assessment_pss_four_item_going_your_way_enum  \\\n",
       "0                                              1.0   \n",
       "1                                              2.0   \n",
       "2                                              0.0   \n",
       "3                                              3.0   \n",
       "4                                              1.0   \n",
       "...                                            ...   \n",
       "5074                                           0.0   \n",
       "5076                                           1.0   \n",
       "5077                                           1.0   \n",
       "5078                                           2.0   \n",
       "5079                                           2.0   \n",
       "\n",
       "      assessment_pss_four_item_handle_problems_enum  \\\n",
       "0                                               3.0   \n",
       "1                                               0.0   \n",
       "2                                               1.0   \n",
       "3                                               2.0   \n",
       "4                                               1.0   \n",
       "...                                             ...   \n",
       "5074                                            0.0   \n",
       "5076                                            2.0   \n",
       "5077                                            0.0   \n",
       "5078                                            1.0   \n",
       "5079                                            1.0   \n",
       "\n",
       "      assessment_pss_four_item_insurmountable_enum  \\\n",
       "0                                              2.0   \n",
       "1                                              1.0   \n",
       "2                                              0.0   \n",
       "3                                              2.0   \n",
       "4                                              1.0   \n",
       "...                                            ...   \n",
       "5074                                           1.0   \n",
       "5076                                           2.0   \n",
       "5077                                           0.0   \n",
       "5078                                           2.0   \n",
       "5079                                           2.0   \n",
       "\n",
       "      assessment_pss_four_item_out_of_control_enum  \n",
       "0                                              2.0  \n",
       "1                                              1.0  \n",
       "2                                              3.0  \n",
       "3                                              3.0  \n",
       "4                                              2.0  \n",
       "...                                            ...  \n",
       "5074                                           0.0  \n",
       "5076                                           2.0  \n",
       "5077                                           0.0  \n",
       "5078                                           3.0  \n",
       "5079                                           3.0  \n",
       "\n",
       "[3616 rows x 3291 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "qs = get_snapshot(\"assessments\", clean=True).sort_values(by=\"days_in_program\")\n",
    "q_features = qs.columns[8:]\n",
    "q_features = q_features[q_features.str.endswith(\"_enum\") | q_features.str.endswith(\"_int\")]\n",
    "qs.days_in_program = qs.days_in_program.astype(\"float64\")\n",
    "\n",
    "# Merge with measure\n",
    "qs_merged = pd.merge_asof(\n",
    "    measure, qs.dropna(subset=\"days_in_program\"), \n",
    "    by=\"public_client_id\", \n",
    "    on=\"days_in_program\", \n",
    "    direction=\"nearest\", \n",
    "    tolerance=30.0).dropna(subset=q_features, how=\"all\")\n",
    "q_features = qs.columns[qs.columns.isin(q_features)]\n",
    "\n",
    "# Filter uncomplete\n",
    "bad = qs_merged[q_features].isnull().sum() / qs_merged.shape[0] > 0.25\n",
    "q_features = bad[~bad].index\n",
    "qs_merged = qs_merged.drop(columns=bad[bad].index)\n",
    "\n",
    "# The data is all categorical so we will make it ordinal now\n",
    "qs_merged[q_features] = qs_merged[q_features].apply(lambda x: x.str.split(\"\\\\(|\\\\)\", regex=True).str[1]).astype(\"float64\")\n",
    "qs_merged"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "407c9a8a-4d10-422b-bebf-d982d1c5d9c6",
   "metadata": {},
   "source": [
    "And we test again."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "0b1982bd-f570-4c9a-bcb3-213c2a88d3ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a7c48da5a9514c7991d284bf5c4d0562",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Output()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"></pre>\n"
      ],
      "text/plain": []
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "args = q_features\n",
    "df = qs_merged  # don't forget to reassign those\n",
    "results = map(get_results, track(args))\n",
    "results = list(results)\n",
    "tests_qs = pd.concat(results)\n",
    "tests_qs[\"q\"] = multipletests(tests_qs.p, method=\"fdr_bh\")[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a021c813-5b15-45dc-aa03-105d51a5cb7a",
   "metadata": {},
   "source": [
    "For the annotations we will simply append the meaning of the numeric scale here. Makes it easier to interpret the result (like the directon of beta)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f6f5f20a-a283-4ef2-b448-4f0165e46c4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "scales = qs[q_features].apply(lambda x: \" | \".join(x.astype(str).sort_values().unique()) if x.dtype == \"O\" else \"numeric\")\n",
    "tests_qs[\"scales\"] = scales[tests_qs.feature].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f1a80481-97df-4fa2-8355-d4ad4d7ddd27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>beta</th>\n",
       "      <th>t_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>n</th>\n",
       "      <th>q</th>\n",
       "      <th>scales</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>assessment_digestion_diarrhea_enum</th>\n",
       "      <td>assessment_digestion_diarrhea_enum</td>\n",
       "      <td>0.247584</td>\n",
       "      <td>8.087418</td>\n",
       "      <td>8.563850e-16</td>\n",
       "      <td>3202.0</td>\n",
       "      <td>7.193634e-14</td>\n",
       "      <td>(1) Daily | (2) Several times per week | (3) O...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assessment_digestion_bowel_movements_enum</th>\n",
       "      <td>assessment_digestion_bowel_movements_enum</td>\n",
       "      <td>-0.174572</td>\n",
       "      <td>-7.288794</td>\n",
       "      <td>4.083475e-13</td>\n",
       "      <td>2715.0</td>\n",
       "      <td>1.715059e-11</td>\n",
       "      <td>(1) 2 or fewer times per week | (2) 3-6 times ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assessment_lifestyle_vigorous_activity_duration_enum</th>\n",
       "      <td>assessment_lifestyle_vigorous_activity_duratio...</td>\n",
       "      <td>0.321142</td>\n",
       "      <td>4.187388</td>\n",
       "      <td>2.898410e-05</td>\n",
       "      <td>3186.0</td>\n",
       "      <td>8.115547e-04</td>\n",
       "      <td>(0) N/A | (1) 10 min | (2) 20 min | (3) 30 min...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assessment_satisfaction_not_healthy_enum</th>\n",
       "      <td>assessment_satisfaction_not_healthy_enum</td>\n",
       "      <td>0.239992</td>\n",
       "      <td>3.995797</td>\n",
       "      <td>6.601961e-05</td>\n",
       "      <td>3022.0</td>\n",
       "      <td>1.386412e-03</td>\n",
       "      <td>(1) Strongly Agree | (2) Moderately Agree | (3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assessment_digestion_nausea_enum</th>\n",
       "      <td>assessment_digestion_nausea_enum</td>\n",
       "      <td>0.077932</td>\n",
       "      <td>3.816247</td>\n",
       "      <td>1.380636e-04</td>\n",
       "      <td>3202.0</td>\n",
       "      <td>2.319468e-03</td>\n",
       "      <td>(1) Daily | (2) Regularly (several times per w...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assessment_satisfaction_world_is_a_bad_place_enum</th>\n",
       "      <td>assessment_satisfaction_world_is_a_bad_place_enum</td>\n",
       "      <td>0.205586</td>\n",
       "      <td>3.729297</td>\n",
       "      <td>1.955416e-04</td>\n",
       "      <td>3027.0</td>\n",
       "      <td>2.737582e-03</td>\n",
       "      <td>(1) Strongly Agree | (2) Moderately Agree | (3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assessment_satisfaction_life_is_rewarding_enum</th>\n",
       "      <td>assessment_satisfaction_life_is_rewarding_enum</td>\n",
       "      <td>0.149907</td>\n",
       "      <td>3.492221</td>\n",
       "      <td>4.858941e-04</td>\n",
       "      <td>3026.0</td>\n",
       "      <td>3.401258e-03</td>\n",
       "      <td>(1) Strongly Disagree | (2) Moderately Disagre...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assessment_satisfaction_have_energy_enum</th>\n",
       "      <td>assessment_satisfaction_have_energy_enum</td>\n",
       "      <td>0.224543</td>\n",
       "      <td>3.631053</td>\n",
       "      <td>2.869626e-04</td>\n",
       "      <td>3022.0</td>\n",
       "      <td>3.401258e-03</td>\n",
       "      <td>(1) Strongly Disagree | (2) Moderately Disagre...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assessment_satisfaction_not_optimistic_enum</th>\n",
       "      <td>assessment_satisfaction_not_optimistic_enum</td>\n",
       "      <td>0.199290</td>\n",
       "      <td>3.528573</td>\n",
       "      <td>4.240383e-04</td>\n",
       "      <td>3026.0</td>\n",
       "      <td>3.401258e-03</td>\n",
       "      <td>(1) Strongly Agree | (2) Moderately Agree | (3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assessment_lifestyle_childer_enum</th>\n",
       "      <td>assessment_lifestyle_childer_enum</td>\n",
       "      <td>0.169065</td>\n",
       "      <td>3.564897</td>\n",
       "      <td>3.693596e-04</td>\n",
       "      <td>3184.0</td>\n",
       "      <td>3.401258e-03</td>\n",
       "      <td>(0) None | (1) One | (2) Two | (3) Three | (4)...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                              feature  \\\n",
       "assessment_digestion_diarrhea_enum                                 assessment_digestion_diarrhea_enum   \n",
       "assessment_digestion_bowel_movements_enum                   assessment_digestion_bowel_movements_enum   \n",
       "assessment_lifestyle_vigorous_activity_duration...  assessment_lifestyle_vigorous_activity_duratio...   \n",
       "assessment_satisfaction_not_healthy_enum                     assessment_satisfaction_not_healthy_enum   \n",
       "assessment_digestion_nausea_enum                                     assessment_digestion_nausea_enum   \n",
       "assessment_satisfaction_world_is_a_bad_place_enum   assessment_satisfaction_world_is_a_bad_place_enum   \n",
       "assessment_satisfaction_life_is_rewarding_enum         assessment_satisfaction_life_is_rewarding_enum   \n",
       "assessment_satisfaction_have_energy_enum                     assessment_satisfaction_have_energy_enum   \n",
       "assessment_satisfaction_not_optimistic_enum               assessment_satisfaction_not_optimistic_enum   \n",
       "assessment_lifestyle_childer_enum                                   assessment_lifestyle_childer_enum   \n",
       "\n",
       "                                                        beta  t_statistic  \\\n",
       "assessment_digestion_diarrhea_enum                  0.247584     8.087418   \n",
       "assessment_digestion_bowel_movements_enum          -0.174572    -7.288794   \n",
       "assessment_lifestyle_vigorous_activity_duration...  0.321142     4.187388   \n",
       "assessment_satisfaction_not_healthy_enum            0.239992     3.995797   \n",
       "assessment_digestion_nausea_enum                    0.077932     3.816247   \n",
       "assessment_satisfaction_world_is_a_bad_place_enum   0.205586     3.729297   \n",
       "assessment_satisfaction_life_is_rewarding_enum      0.149907     3.492221   \n",
       "assessment_satisfaction_have_energy_enum            0.224543     3.631053   \n",
       "assessment_satisfaction_not_optimistic_enum         0.199290     3.528573   \n",
       "assessment_lifestyle_childer_enum                   0.169065     3.564897   \n",
       "\n",
       "                                                               p       n  \\\n",
       "assessment_digestion_diarrhea_enum                  8.563850e-16  3202.0   \n",
       "assessment_digestion_bowel_movements_enum           4.083475e-13  2715.0   \n",
       "assessment_lifestyle_vigorous_activity_duration...  2.898410e-05  3186.0   \n",
       "assessment_satisfaction_not_healthy_enum            6.601961e-05  3022.0   \n",
       "assessment_digestion_nausea_enum                    1.380636e-04  3202.0   \n",
       "assessment_satisfaction_world_is_a_bad_place_enum   1.955416e-04  3027.0   \n",
       "assessment_satisfaction_life_is_rewarding_enum      4.858941e-04  3026.0   \n",
       "assessment_satisfaction_have_energy_enum            2.869626e-04  3022.0   \n",
       "assessment_satisfaction_not_optimistic_enum         4.240383e-04  3026.0   \n",
       "assessment_lifestyle_childer_enum                   3.693596e-04  3184.0   \n",
       "\n",
       "                                                               q  \\\n",
       "assessment_digestion_diarrhea_enum                  7.193634e-14   \n",
       "assessment_digestion_bowel_movements_enum           1.715059e-11   \n",
       "assessment_lifestyle_vigorous_activity_duration...  8.115547e-04   \n",
       "assessment_satisfaction_not_healthy_enum            1.386412e-03   \n",
       "assessment_digestion_nausea_enum                    2.319468e-03   \n",
       "assessment_satisfaction_world_is_a_bad_place_enum   2.737582e-03   \n",
       "assessment_satisfaction_life_is_rewarding_enum      3.401258e-03   \n",
       "assessment_satisfaction_have_energy_enum            3.401258e-03   \n",
       "assessment_satisfaction_not_optimistic_enum         3.401258e-03   \n",
       "assessment_lifestyle_childer_enum                   3.401258e-03   \n",
       "\n",
       "                                                                                               scales  \n",
       "assessment_digestion_diarrhea_enum                  (1) Daily | (2) Several times per week | (3) O...  \n",
       "assessment_digestion_bowel_movements_enum           (1) 2 or fewer times per week | (2) 3-6 times ...  \n",
       "assessment_lifestyle_vigorous_activity_duration...  (0) N/A | (1) 10 min | (2) 20 min | (3) 30 min...  \n",
       "assessment_satisfaction_not_healthy_enum            (1) Strongly Agree | (2) Moderately Agree | (3...  \n",
       "assessment_digestion_nausea_enum                    (1) Daily | (2) Regularly (several times per w...  \n",
       "assessment_satisfaction_world_is_a_bad_place_enum   (1) Strongly Agree | (2) Moderately Agree | (3...  \n",
       "assessment_satisfaction_life_is_rewarding_enum      (1) Strongly Disagree | (2) Moderately Disagre...  \n",
       "assessment_satisfaction_have_energy_enum            (1) Strongly Disagree | (2) Moderately Disagre...  \n",
       "assessment_satisfaction_not_optimistic_enum         (1) Strongly Agree | (2) Moderately Agree | (3...  \n",
       "assessment_lifestyle_childer_enum                   (0) None | (1) One | (2) Two | (3) Three | (4)...  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tests_qs.sort_values(by=\"q\").head(10)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
